// tnp-ui/errors-notyfications/errors-notyfications.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var tnp_helpers_1 = require("tnp-helpers");
var errors_notyfications_service_1 = require("./errors-notyfications.service");
var ErrorsNotyficationsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ErrorsNotyficationsComponent, _super);
    function ErrorsNotyficationsComponent(notificaitonService) {
        var _this = _super.call(this) || this;
        _this.notificaitonService = notificaitonService;
        return _this;
    }
    ErrorsNotyficationsComponent.prototype.ngOnInit = function () {
    };
    ErrorsNotyficationsComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.notificaitonService.init(_this.handlers, _this.template, (function (data) {
                _this.data = data;
            }));
        });
    };
    tslib_1.__decorate([
        core_1.ViewChild('tmp'),
        tslib_1.__metadata("design:type", core_1.TemplateRef)
    ], ErrorsNotyficationsComponent.prototype, "template", void 0);
    ErrorsNotyficationsComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-errors-notyfications',
            templateUrl: './errors-notyfications.component.html',
            styleUrls: ['./errors-notyfications.component.scss']
        }),
        tslib_1.__metadata("design:paramtypes", [errors_notyfications_service_1.ErrorsNotyficationsService])
    ], ErrorsNotyficationsComponent);
    return ErrorsNotyficationsComponent;
}(tnp_helpers_1.BaseComponent));
exports.ErrorsNotyficationsComponent = ErrorsNotyficationsComponent;
