// tnp-ui/formly-switch/formly-switch.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
// formly
var core_2 = require("@ngx-formly/core");
// other
var _ = require("lodash");
var ng2_logger_1 = require("ng2-logger");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var log = ng2_logger_1.Log.create('formly switch');
// @Morphi.Formly.RegisterComponentAsType('FormlySwitchComponent', 'switch')
var FormlySwitchComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormlySwitchComponent, _super);
    function FormlySwitchComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FormlySwitchComponent.prototype.change = function (e) {
        this.formControl.setValue(e.checked);
    };
    FormlySwitchComponent.prototype.ngOnInit = function () {
        if (_.isUndefined(this.field.defaultValue)) {
            this.field.defaultValue = false;
        }
        if (!this.field.templateOptions.label) {
            this.field.templateOptions.label = 'button with action';
        }
    };
    FormlySwitchComponent = tslib_1.__decorate([
        typescript_class_helpers_1.CLASS.NAME('FormlySwitchComponent'),
        core_1.Component({
            selector: 'app-formly-switch',
            templateUrl: './formly-switch.component.html',
            styleUrls: ['./formly-switch.component.scss']
        })
    ], FormlySwitchComponent);
    return FormlySwitchComponent;
}(core_2.FieldType));
exports.FormlySwitchComponent = FormlySwitchComponent;
