// tnp-ui/formly-switch/formly-switch.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
// formlu
var formly_switch_component_1 = require("./formly-switch.component");
var core_2 = require("@ngx-formly/core");
// import { FormlyMaterialModule, } from '@ngx-formly/material';
// material
var slide_toggle_1 = require("@angular/material/slide-toggle");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var angularModules = [
    common_1.CommonModule,
    forms_1.ReactiveFormsModule
];
var formlyModules = [
// FormlyMaterialModule,
];
var materialModules = [
    slide_toggle_1.MatSlideToggleModule
];
exports.FromlModuleMode = core_2.FormlyModule.forRoot({
    types: [
        { name: 'switch', component: formly_switch_component_1.FormlySwitchComponent },
        { name: typescript_class_helpers_1.CLASS.getName(formly_switch_component_1.FormlySwitchComponent), component: formly_switch_component_1.FormlySwitchComponent }
    ]
});
var FormlySwitchModule = /** @class */ (function () {
    function FormlySwitchModule() {
    }
    FormlySwitchModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays(angularModules, formlyModules, materialModules, [
                exports.FromlModuleMode,
            ]),
            exports: [
                formly_switch_component_1.FormlySwitchComponent,
            ],
            declarations: [formly_switch_component_1.FormlySwitchComponent]
        })
    ], FormlySwitchModule);
    return FormlySwitchModule;
}());
exports.FormlySwitchModule = FormlySwitchModule;
