// tnp-helpers/helpers-path.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tnp_core_1 = require("tnp-core");
var tnp_config_1 = require("tnp-config");
var constants_1 = require("./constants");
var HelpersPath = /** @class */ (function () {
    function HelpersPath() {
    }
    HelpersPath.prototype.create = function () {
        var pathPart = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            pathPart[_i] = arguments[_i];
        }
        return tnp_core_1.path.join.apply(tnp_core_1.path, pathPart);
    };
    HelpersPath.prototype.PREFIX = function (baseFileName) {
        return "" + constants_1.BaselineSiteJoinprefix + baseFileName;
    };
    HelpersPath.prototype.removeRootFolder = function (filePath) {
        return filePath.replace(new RegExp("^" + tnp_config_1.config.regexString.pathPartStringRegex, 'g'), '');
    };
    HelpersPath.prototype.removeExtension = function (filePath) {
        var ext = tnp_core_1.path.extname(filePath);
        return tnp_core_1.path.join(tnp_core_1.path.dirname(filePath), tnp_core_1.path.basename(filePath, ext));
    };
    HelpersPath.prototype.removeExt = function (filePath) {
        return filePath.replace(/\.[^\/.]+$/, '');
    };
    return HelpersPath;
}());
exports.HelpersPath = HelpersPath;
