// tnp-helpers/helpers-array-obj.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tnp_core_1 = require("tnp-core");
var fuzzy = require("fuzzy");
var index_1 = require("./index");
var HelpersArrayObj = /** @class */ (function () {
    function HelpersArrayObj() {
    }
    HelpersArrayObj.prototype.from = function (s) {
        if (tnp_core_1._.isArray(s)) {
            return s;
        }
        if (tnp_core_1._.isString(s)) {
            return s.split(' ');
        }
    };
    HelpersArrayObj.prototype.arrayMoveElementBefore = function (arr, a, b, prop) {
        var indexA = prop ? arr.findIndex(function (elem) { return elem[prop] === a[prop]; }) : arr.indexOf(a);
        tnp_core_1._.pullAt(arr, indexA);
        var indexB = prop ? arr.findIndex(function (elem) { return elem[prop] === b[prop]; }) : arr.indexOf(b);
        if (indexB === 0) {
            arr.unshift(a);
        }
        else {
            arr.splice(indexB - 1, 0, a);
        }
        return arr;
    };
    HelpersArrayObj.prototype.arrayMoveElementAfterB = function (arr, a, b, prop) {
        var indexA = prop ? arr.findIndex(function (elem) { return elem[prop] === a[prop]; }) : arr.indexOf(a);
        tnp_core_1._.pullAt(arr, indexA);
        var indexB = prop ? arr.findIndex(function (elem) { return elem[prop] === b[prop]; }) : arr.indexOf(b);
        if (indexB === arr.length - 1) {
            arr.push(a);
        }
        else {
            arr.splice(indexB + 1, 0, a);
        }
        return arr;
    };
    HelpersArrayObj.prototype.uniqArray = function (array, uniqueProperty) {
        var seen = {};
        return array
            .filter(function (f) { return !!f; })
            .filter(function (item) {
            return seen.hasOwnProperty(uniqueProperty ? item[uniqueProperty] : item) ? false
                : (seen[uniqueProperty ? item[uniqueProperty] : item] = true);
        });
    };
    HelpersArrayObj.prototype.sortKeys = function (obj) {
        if (tnp_core_1._.isArray(obj)) {
            return obj.map(index_1.Helpers.arrays.sortKeys);
        }
        if (tnp_core_1._.isObject(obj)) {
            return tnp_core_1._.fromPairs(tnp_core_1._.keys(obj).sort().map(function (key) { return [key, index_1.Helpers.arrays.sortKeys(obj[key])]; }));
        }
        return obj;
    };
    ;
    /**
     * Fuzzy search
     */
    HelpersArrayObj.prototype.fuzzy = function (query, list, valueFn) {
        var resultsFuzzy = fuzzy.filter(query, list.map(function (k) { return valueFn ? valueFn(k) : k; }));
        var resultsFuzzyKebab = fuzzy.filter(tnp_core_1._.kebabCase(query), list.map(function (k) { return tnp_core_1._.kebabCase((valueFn ? valueFn(k) : k)); }));
        var matches = resultsFuzzy.map(function (el) { return el.string; });
        var matchesKebab = resultsFuzzyKebab.map(function (el) { return el.string; });
        var results = (resultsFuzzy.length === 0) ? [] : list.filter(function (k) {
            return matches.includes((valueFn ? valueFn(k) : k));
        });
        if (matches.length === 0 && matchesKebab.length > 0) {
            var m = list.find(function (k) { return tnp_core_1._.kebabCase((valueFn ? valueFn(k) : k)) === tnp_core_1._.first(matchesKebab); });
            results.push(m);
            matches.push((valueFn ? valueFn(m) : m));
        }
        return { matches: matches, results: results };
    };
    return HelpersArrayObj;
}());
exports.HelpersArrayObj = HelpersArrayObj;
