// tnp-ui/notifications/notifications.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _ = require("lodash");
var core_1 = require("@angular/core");
var ngx_toastr_1 = require("ngx-toastr");
var notification_model_1 = require("./notification-model");
var NotificationsService = /** @class */ (function () {
    function NotificationsService(toast) {
        this.toast = toast;
        this.config = {
            closeButton: true,
            easeTime: 5500
        };
    }
    NotificationsService.prototype.meta = function (type, options) {
        if (_.isString(options)) {
            options = {
                title: options
            };
        }
        var title = options.title, subtitle = options.subtitle;
        var t = this.toast[type](title, subtitle, this.config);
        return new notification_model_1.NotificaitonModel(t);
    };
    NotificationsService.prototype.success = function (options) {
        return this.meta('success', options);
    };
    NotificationsService.prototype.error = function (options) {
        return this.meta('error', options);
    };
    NotificationsService.prototype.warn = function (options) {
        return this.meta('warning', options);
    };
    NotificationsService.prototype.info = function (options) {
        return this.meta('info', options);
    };
    NotificationsService = tslib_1.__decorate([
        core_1.Injectable(),
        tslib_1.__metadata("design:paramtypes", [ngx_toastr_1.ToastrService])
    ], NotificationsService);
    return NotificationsService;
}());
exports.NotificationsService = NotificationsService;
