// tnp/apps/project/project-ui/tnp-project/items/item-serve/item-serve.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
//#region angular
var core_1 = require("@angular/core");
//#endregion
//#region isomorphic
var base_item_stepper_1 = require("../base-item-stepper");
var tabs_menu_tnp_project_1 = require("../../tabs-menu-tnp-project");
//#endregion
var ItemServeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ItemServeComponent, _super);
    function ItemServeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ItemServeComponent.prototype.tabSelectedAction = function () {
    };
    ItemServeComponent.prototype.formValueChanged = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    ItemServeComponent.prototype.tabNumber = function () {
        return tabs_menu_tnp_project_1.TnpProjectTabIndex.SERVE;
    };
    ItemServeComponent.prototype.ngOnInit = function () {
    };
    ItemServeComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-item-serve',
            templateUrl: './item-serve.component.html',
            styleUrls: ['./item-serve.component.scss']
        })
    ], ItemServeComponent);
    return ItemServeComponent;
}(base_item_stepper_1.BaseItemStepperProcessBuildComponent));
exports.ItemServeComponent = ItemServeComponent;
