// typescript-class-helpers/tests/classname.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var mocha_1 = require("mocha");
var chai_1 = require("chai");
var index_1 = require("../index");
var classname_1 = require("../classname");
var Proj = /** @class */ (function () {
    function Proj() {
        this.isProjectInstance = true;
    }
    Proj = tslib_1.__decorate([
        index_1.CLASS.NAME('Proj', {
            uniqueKey: 'location'
        })
    ], Proj);
    return Proj;
}());
exports.Proj = Proj;
var Proj2 = /** @class */ (function (_super) {
    tslib_1.__extends(Proj2, _super);
    function Proj2() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isProjectInstance2 = true;
        return _this;
    }
    Proj2 = tslib_1.__decorate([
        index_1.CLASS.NAME('Proj2')
    ], Proj2);
    return Proj2;
}(Proj));
exports.Proj2 = Proj2;
var Proj3 = /** @class */ (function (_super) {
    tslib_1.__extends(Proj3, _super);
    function Proj3() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isProjectInstance2 = true;
        return _this;
    }
    Proj3 = tslib_1.__decorate([
        index_1.CLASS.NAME('Proj3', {
            uniqueKey: 'idaaa',
            classFamily: 'aaaa'
        })
    ], Proj3);
    return Proj3;
}(Proj2));
exports.Proj3 = Proj3;
var P4 = /** @class */ (function (_super) {
    tslib_1.__extends(P4, _super);
    function P4() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return P4;
}(Proj3));
exports.P4 = P4;
var P5 = /** @class */ (function (_super) {
    tslib_1.__extends(P5, _super);
    function P5() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return P5;
}(P4));
exports.P5 = P5;
// const instance = BrowserDB.instance;
mocha_1.describe('CLASSNAME', function () {
    mocha_1.it('Family name shoudl work', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            chai_1.expect(classname_1.CLASSNAME.getClassFamilyByClassName('Proj')).to.be.eq('Proj');
            chai_1.expect(classname_1.CLASSNAME.getClassFamilyByClassName('Proj2')).to.be.eq('Proj');
            chai_1.expect(classname_1.CLASSNAME.getClassFamilyByClassName('Proj3')).to.be.eq('aaaa');
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('Unique keys shoudl work', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            chai_1.expect(index_1.CLASS.OBJECT(new Proj()).indexProperty).to.be.eq('location');
            chai_1.expect(index_1.CLASS.OBJECT(new Proj2()).indexProperty).to.be.eq('id');
            chai_1.expect(index_1.CLASS.OBJECT(new Proj3()).indexProperty).to.be.eq('idaaa');
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('Name inheritance should works', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            chai_1.expect(index_1.CLASS.getName(P4)).to.be.eq('Proj3');
            chai_1.expect(index_1.CLASS.getName(P5)).to.be.eq('Proj3');
            chai_1.expect(index_1.CLASS.getName(P4, true)).to.be.eq('Proj3');
            chai_1.expect(index_1.CLASS.getName(P5, true)).to.be.eq('Proj3');
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('Should handle production mode getClassName', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var prev, called;
        return tslib_1.__generator(this, function (_a) {
            chai_1.expect(index_1.CLASS.getName(Object)).to.be.eq('Object');
            prev = console.error;
            called = false;
            console.error = function (arg) {
                called = true;
            };
            chai_1.expect(index_1.CLASS.getName(Object, true)).to.be.eq('Object');
            console.error = prev;
            chai_1.expect(called).to.be.true;
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('Should handle class object propertyly', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            chai_1.expect(index_1.CLASS.OBJECT({}).isClassObject).to.be.eq(false);
            chai_1.expect(index_1.CLASS.OBJECT(new P4()).isClassObject).to.be.eq(true);
            chai_1.expect(index_1.CLASS.OBJECT(new P5()).isClassObject).to.be.eq(true);
            chai_1.expect(index_1.CLASS.OBJECT(new Date()).isClassObject).to.be.eq(true);
            chai_1.expect(index_1.CLASS.OBJECT([]).isClassObject).to.be.eq(false);
            chai_1.expect(index_1.CLASS.OBJECT(true).isClassObject).to.be.eq(false);
            chai_1.expect(index_1.CLASS.OBJECT(100).isClassObject).to.be.eq(false);
            chai_1.expect(index_1.CLASS.OBJECT(/asd/g).isClassObject).to.be.eq(false);
            return [2 /*return*/];
        });
    }); });
});
