// navi-cli/scripts/GEN.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _ = require("lodash");
var path = require("path");
var base_generators_backend_1 = require("../generator/base-generators.backend");
function GEN(files, type, decoratorFileName) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var cwdFromArgsIndex, cwdFromArgs, cwd, index, cmpPathRelative, location_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    files = _.cloneDeep(files);
                    cwdFromArgsIndex = files.findIndex(function (f, i) { return f === '--cwd'; });
                    cwdFromArgs = (cwdFromArgsIndex !== -1) ? files[cwdFromArgsIndex + 1] : void 0;
                    cwd = _.isString(cwdFromArgs) ? cwdFromArgs : process.cwd();
                    if (cwdFromArgsIndex !== -1) {
                        files = files.filter(function (f, i) {
                            if (f === '--cwd') {
                                files[i + 1] = '';
                                return false;
                            }
                            return true;
                        })
                            .filter(function (f) { return !!f; });
                    }
                    if (!_.isArray(files)) return [3 /*break*/, 4];
                    index = 0;
                    _a.label = 1;
                case 1:
                    if (!(index < files.length)) return [3 /*break*/, 4];
                    cmpPathRelative = files[index];
                    location_1 = cmpPathRelative;
                    if (!path.isAbsolute(location_1)) {
                        location_1 = path.join(cwd, cmpPathRelative);
                    }
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(location_1, type, _.isString(decoratorFileName) && decoratorFileName.trim() !== '', decoratorFileName)];
                case 2:
                    (_a.sent()).save();
                    _a.label = 3;
                case 3:
                    index++;
                    return [3 /*break*/, 1];
                case 4:
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_COMPONENT(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'component')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_MODULE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'module')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_ROUTING_MODULE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'routing:module')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_DIRECTIVE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'directive')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_PIPE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'pipe')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_SERVICE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'service')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_API_SERVICE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, GEN(args, 'api:service')];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_EXAMPLE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var cmpPathRelative;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    cmpPathRelative = _.first(args);
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(path.join(process.cwd(), cmpPathRelative), 'service')];
                case 1:
                    (_a.sent()).save();
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(path.join(process.cwd(), cmpPathRelative), 'component')];
                case 2:
                    (_a.sent()).save();
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(path.join(process.cwd(), cmpPathRelative), 'module')];
                case 3:
                    (_a.sent()).save();
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_API_MODULE(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var cmpPathRelative;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    cmpPathRelative = _.first(args);
                    if (cmpPathRelative.endsWith('/')) {
                        cmpPathRelative = cmpPathRelative.slice(0, cmpPathRelative.length - 1);
                    }
                    if (cmpPathRelative.endsWith('-api')) {
                        cmpPathRelative = cmpPathRelative.replace('-api', '');
                    }
                    cmpPathRelative = cmpPathRelative + "/" + path.basename(cmpPathRelative) + "-api";
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(path.join(process.cwd(), cmpPathRelative), 'api:service')];
                case 1:
                    (_a.sent()).save();
                    return [4 /*yield*/, base_generators_backend_1.FilesGenerator.From(path.join(process.cwd(), cmpPathRelative), 'module')];
                case 2:
                    (_a.sent()).save();
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function GEN_DECORATOR(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var _a, type, oldfilename;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = require('minimist')(process.argv), type = _a.type, oldfilename = _a.oldfilename;
                    return [4 /*yield*/, GEN(args.filter(function (a) { return !a.startsWith('--type') && !a.startsWith('--oldfilename'); }), type, oldfilename)];
                case 1:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.default = {
    GEN_COMPONENT: GEN_COMPONENT,
    GEN_MODULE: GEN_MODULE,
    GEN_ROUTING_MODULE: GEN_ROUTING_MODULE,
    GEN_API_MODULE: GEN_API_MODULE,
    GEN_DIRECTIVE: GEN_DIRECTIVE,
    GEN_PIPE: GEN_PIPE,
    GEN_SERVICE: GEN_SERVICE,
    GEN_API_SERVICE: GEN_API_SERVICE,
    GEN_EXAMPLE: GEN_EXAMPLE,
    GEN_DECORATOR: GEN_DECORATOR
};
