// navi-cli/scripts/GIT.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RESETHARD = exports.RESETSOFT = exports.PULL = exports.REBASE = exports.COMMIT = exports.PUSHFORCE = exports.PUSH = void 0;
var project_backend_1 = require("../project.backend");
function PUSH() {
    project_backend_1.Project.Current.git.pushCurrentBranch(false);
    process.exit(0);
}
exports.PUSH = PUSH;
function PUSHFORCE() {
    project_backend_1.Project.Current.git.pushCurrentBranch(true);
    process.exit(0);
}
exports.PUSHFORCE = PUSHFORCE;
function COMMIT(args) {
    project_backend_1.Project.Current.git.commit(args.join(' '));
    process.exit(0);
}
exports.COMMIT = COMMIT;
function REBASE(args) {
    project_backend_1.Project.Current.helpers.run("git rebsae " + args.join(' '));
    process.exit(0);
}
exports.REBASE = REBASE;
function PULL(args) {
    project_backend_1.Project.Current.helpers.run("git pull origin " + project_backend_1.Project.Current.git.currentBranchName).sync();
    process.exit(0);
}
exports.PULL = PULL;
function RESETSOFT(args) {
    project_backend_1.Project.Current.helpers.run("git reset --soft HEAD~1").sync();
    process.exit(0);
}
exports.RESETSOFT = RESETSOFT;
function RESETHARD(args) {
    project_backend_1.Project.Current.helpers.run("git reset --hard HEAD~1").sync();
    process.exit(0);
}
exports.RESETHARD = RESETHARD;
exports.default = {
    PUSH: PUSH,
    PUSHFORCE: PUSHFORCE,
    COMMIT: COMMIT,
    REBASE: REBASE,
    PULL: PULL,
    RESETSOFT: RESETSOFT,
    RESETHARD: RESETHARD,
};
