// navi-cli/scripts/START.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REPLAY = exports.RECORDDEFAULT = exports.RECORDLAST = exports.RECORD = exports.STARTDEFAULT = exports.STARTLAST = exports.START = void 0;
var tslib_1 = require("tslib");
var project_backend_1 = require("../project.backend");
function START(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.startProcess()];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.START = START;
function STARTLAST(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.startProcess('start', 'last')];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.STARTLAST = STARTLAST;
function STARTDEFAULT(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.startProcess('start', 'default')];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.STARTDEFAULT = STARTDEFAULT;
function RECORD(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.record(void 0, args)];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.RECORD = RECORD;
function RECORDLAST(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.record('last', args)];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.RECORDLAST = RECORDLAST;
function RECORDDEFAULT(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.record('default', args)];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.RECORDDEFAULT = RECORDDEFAULT;
function REPLAY(exit, args) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.replay()];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.REPLAY = REPLAY;
exports.default = {
    START: START,
    STARTLAST: STARTLAST,
    STARTDEFAULT: STARTDEFAULT,
    RECORD: RECORD,
    RECORDLAST: RECORDLAST,
    RECORDDEFAULT: RECORDDEFAULT,
    REPLAY: REPLAY,
};
