// navi-cli/scripts/VSCODE.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VSCODE = exports.VSCODE_EXT = void 0;
var tslib_1 = require("tslib");
var path = require("path");
var project_backend_1 = require("../project.backend");
var tnp_helpers_1 = require("tnp-helpers");
function VSCODE_EXT(args, exit) {
    if (exit === void 0) { exit = true; }
    tnp_helpers_1.Helpers.run("npm install && tsc && npm run build:install", {
        cwd: path.join(__dirname, '..', '..', 'plugins', 'navi-vscode-ext'),
        output: true
    }).sync();
    if (exit) {
        process.exit(0);
    }
}
exports.VSCODE_EXT = VSCODE_EXT;
function VSCODE(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_backend_1.Project.Current.vscode.update()];
                case 1:
                    _a.sent();
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.VSCODE = VSCODE;
exports.default = {
    VSCODE: VSCODE,
    VSCODE_EXT: VSCODE_EXT,
    EXTENSION: function (args) {
        VSCODE_EXT(args);
    },
    VSCODE_EXTENSION: function (args) {
        VSCODE_EXT(args);
    },
    EXT: function (args) {
        VSCODE_EXT(args);
    }
};
