// tnp-ui/form/form-wrapper-material.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// angular
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
// material
var button_1 = require("@angular/material/button");
var icon_1 = require("@angular/material/icon");
var slide_toggle_1 = require("@angular/material/slide-toggle");
var dialog_1 = require("@angular/material/dialog");
var progress_bar_1 = require("@angular/material/progress-bar");
// formly
var core_2 = require("@ngx-formly/core");
var input_1 = require("@ngx-formly/material/input");
var toggle_1 = require("@ngx-formly/material/toggle");
var datepicker_1 = require("@ngx-formly/material/datepicker");
var core_3 = require("@angular/material/core");
var slider_1 = require("@ngx-formly/material/slider");
var morphi_1 = require("morphi");
var morphi_2 = require("morphi");
// custom formly components
var select_1 = require("../select");
// other
var ngx_pipes_1 = require("ngx-pipes");
// local
var form_wrapper_material_component_1 = require("./form-wrapper-material.component");
// base components
var editor_1 = require("../editor");
// aditional types componets
var additional_types_1 = require("./additional-types");
var formly_buttion_with_action_1 = require("../formly-buttion-with-action");
// import { ProcessLoggerModule } from '../entity-components/process-logger';
exports.FormlyModuleMod = core_2.FormlyModule.forRoot({
    types: [
        // { name: 'switch', component: FormlySwitchComponent },
        { name: 'iconbutton', component: additional_types_1.IconButtonWithActionComponent },
        { name: 'repeat', component: morphi_1.RepeatTypeComponent }
    ],
    validationMessages: [
        { name: 'required', message: 'This field is required' },
    ],
    wrappers: [{ name: 'groupwrap', component: morphi_2.FormlyHorizontalWrapper }],
});
var angularModules = [
    common_1.CommonModule,
    forms_1.ReactiveFormsModule,
];
var materialModules = [
    button_1.MatButtonModule,
    icon_1.MatIconModule,
    dialog_1.MatDialogModule,
    toggle_1.FormlyMatToggleModule,
    slide_toggle_1.MatSlideToggleModule,
    progress_bar_1.MatProgressBarModule
];
var myFormlyModules = [
    select_1.SelectWrapperModule,
    editor_1.EditorWrapperModule,
    formly_buttion_with_action_1.FormlyButtionWithActionModule,
];
var formlyModules = [
    // FormlyMaterialModule,
    exports.FormlyModuleMod,
    input_1.FormlyMatInputModule,
    toggle_1.FormlyMatToggleModule,
    datepicker_1.FormlyMatDatepickerModule,
    core_3.MatNativeDateModule,
    slider_1.FormlyMatSliderModule,
    // custom
    ngx_pipes_1.NgStringPipesModule
];
var customComponetns = [
    form_wrapper_material_component_1.FormWrapperMaterialComponent,
    additional_types_1.IconButtonWithActionComponent,
    morphi_1.RepeatTypeComponent,
    morphi_2.FormlyHorizontalWrapper
];
var entityModules = [
// ProcessLoggerModule
];
var FormWrapperMaterialModule = /** @class */ (function () {
    function FormWrapperMaterialModule() {
    }
    FormWrapperMaterialModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays(angularModules, formlyModules, myFormlyModules, materialModules, entityModules),
            exports: tslib_1.__spreadArrays(myFormlyModules, customComponetns),
            declarations: tslib_1.__spreadArrays(customComponetns)
        })
    ], FormWrapperMaterialModule);
    return FormWrapperMaterialModule;
}());
exports.FormWrapperMaterialModule = FormWrapperMaterialModule;
