// tnp/apps/process/process-logger-base.class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tnp_helpers_1 = require("tnp-helpers");
var core_1 = require("@angular/core");
var ProcessLoggerBaseClass = /** @class */ (function (_super) {
    tslib_1.__extends(ProcessLoggerBaseClass, _super);
    function ProcessLoggerBaseClass(elemetRef) {
        var _this = _super.call(this) || this;
        _this.elemetRef = elemetRef;
        return _this;
    }
    // @Input() @HostBinding('style.height.px') height = 190;
    ProcessLoggerBaseClass.prototype.ngAfterViewInit = function () {
        var _this = this;
        // this.messages = this.model.allProgressData;
        setTimeout(function () {
            // this.elemetRef.nativeElement.style.height = `${this.height}px`;
            _this.scrollDown();
        });
    };
    ProcessLoggerBaseClass.prototype.scrollDown = function () {
        var _this = this;
        setTimeout(function () {
            if (_this.elemetRef && _this.elemetRef.nativeElement) {
                _this.elemetRef.nativeElement.scrollTop = _this.elemetRef.nativeElement.scrollHeight;
            }
        }, 100);
    };
    ProcessLoggerBaseClass = tslib_1.__decorate([
        core_1.Component({
            selector: 'process-logger-base',
            template: ''
        }),
        tslib_1.__metadata("design:paramtypes", [core_1.ElementRef])
    ], ProcessLoggerBaseClass);
    return ProcessLoggerBaseClass;
}(tnp_helpers_1.BaseFormlyComponent));
exports.ProcessLoggerBaseClass = ProcessLoggerBaseClass;
