// background-worker-process/base-worker-child-controller.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tnp_core_1 = require("tnp-core");
var morphi_1 = require("morphi");
var base_worker_controller_backend_1 = require("./base-worker-controller.backend");
var bootsrap_worker_backend_1 = require("./bootsrap-worker.backend");
//#region @notForNpm
var test_entity_backend_1 = require("./test-entity.backend");
//#endregion
var typescript_class_helpers_1 = require("typescript-class-helpers");
var BaseWorkerChildController = /** @class */ (function (_super) {
    tslib_1.__extends(BaseWorkerChildController, _super);
    function BaseWorkerChildController() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(BaseWorkerChildController.prototype, "filename", {
        get: function () {
            return tnp_core_1.crossPlatformPath(__filename);
        },
        enumerable: true,
        configurable: true
    });
    BaseWorkerChildController.prototype.html = function () {
        var _this = this;
        return function (req, res) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, "<h1> hello worker </h1>"];
        }); }); };
    };
    BaseWorkerChildController.prototype.hello = function () {
        var _this = this;
        return function (req, res) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, 'hello from child worker!'];
        }); }); };
    };
    BaseWorkerChildController.prototype.hello2 = function () {
        var _this = this;
        return function (req, res) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, 'hello 2222!'];
        }); }); };
    };
    BaseWorkerChildController.prototype.updateRealtime = function () {
        var _this = this;
        //#region @notForNpm
        var id = 2;
        morphi_1.Morphi.Realtime.Server.TrigggerEntityChanges(test_entity_backend_1.TestEntity2.by(id));
        var msg = "realtime update of (" + typescript_class_helpers_1.CLASS.getName(test_entity_backend_1.TestEntity2) + ", id:" + id + ").. from worker " + typescript_class_helpers_1.CLASS.getNameFromObject(this);
        this.updates.push("[" + (new Date).getTime() + "] " + msg);
        setTimeout(function () {
            _this.updateRealtime();
        }, 2000);
        //#endregion
    };
    BaseWorkerChildController.prototype.initExampleDbData = function (isWorker) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                if (isWorker) {
                    setTimeout(function () {
                        _this.updateRealtime();
                    }, 2000);
                }
                return [2 /*return*/];
            });
        });
    };
    tslib_1.__decorate([
        morphi_1.Morphi.Http.GET('/'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", Object)
    ], BaseWorkerChildController.prototype, "html", null);
    tslib_1.__decorate([
        morphi_1.Morphi.Http.GET(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", Object)
    ], BaseWorkerChildController.prototype, "hello", null);
    BaseWorkerChildController = tslib_1.__decorate([
        morphi_1.Morphi.Controller({
            className: 'BaseWorkerChildController',
        })
    ], BaseWorkerChildController);
    return BaseWorkerChildController;
}(base_worker_controller_backend_1.BaseWorkerController));
exports.BaseWorkerChildController = BaseWorkerChildController;
exports.default = bootsrap_worker_backend_1.BootstrapWorker.bootstrap(BaseWorkerChildController);
