// tnp-db/scripts/monit-commands.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tnp_helpers_1 = require("tnp-helpers");
var DBMonitCommands = /** @class */ (function () {
    function DBMonitCommands(db) {
        this.db = db;
        this.counter = 0;
    }
    DBMonitCommands.prototype.start = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.repeat();
                return [2 /*return*/];
            });
        });
    };
    DBMonitCommands.prototype.repeat = function (n) {
        if (n === void 0) { n = 0; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var cmds;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.db.getCommands()];
                    case 1:
                        cmds = _a.sent();
                        tnp_helpers_1.Helpers.clearConsole();
                        tnp_helpers_1.Helpers.info("COMMANDS " + ++this.counter + " :");
                        console.table(cmds);
                        // cmds.forEach(p => {
                        //   Helpers.log(`\t${p.location}\t${p.command}\t${p.shortCommandForLastCommand}\n`);
                        // });
                        setTimeout(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.repeat(n + 1)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); }, 1000);
                        return [2 /*return*/];
                }
            });
        });
    };
    return DBMonitCommands;
}());
exports.DBMonitCommands = DBMonitCommands;
