// tnp-db/entites/process-instance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var base_entity_1 = require("./base-entity");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var ProcessInstance = /** @class */ (function (_super) {
    tslib_1.__extends(ProcessInstance, _super);
    function ProcessInstance() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ProcessInstance.prototype.isEqual = function (anotherInstace) {
        return (this.className === anotherInstace.className &&
            this.entityId === anotherInstace.entityId &&
            this.entityProperty === anotherInstace.entityProperty);
    };
    ProcessInstance.prototype.setInfo = function (metaInfo) {
        var className = metaInfo.className, entityId = metaInfo.entityId, entityProperty = metaInfo.entityProperty;
        this.className = className;
        this.entityId = entityId;
        this.entityProperty = entityProperty;
    };
    Object.defineProperty(ProcessInstance.prototype, "info", {
        get: function () {
            return {
                className: this.className,
                entityId: this.entityId,
                entityProperty: this.entityProperty,
            };
        },
        enumerable: true,
        configurable: true
    });
    ProcessInstance = tslib_1.__decorate([
        typescript_class_helpers_1.CLASS.NAME('ProcessInstance')
    ], ProcessInstance);
    return ProcessInstance;
}(base_entity_1.DBBaseEntity));
exports.ProcessInstance = ProcessInstance;
//#endregion
