// morphi/realtime/realtime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RealtimeBase = /** @class */ (function () {
    function RealtimeBase(context) {
        this.context = context;
    }
    Object.defineProperty(RealtimeBase.prototype, "socketNamespace", {
        //#endregion
        get: function () {
            var self = this;
            return {
                set FE(v) {
                    self.socketFrontEnd = v;
                },
                get FE() {
                    return self.socketFrontEnd;
                },
                set FE_REALTIME(v) {
                    self.socketFrontEndRealtime = v;
                },
                get FE_REALTIME() {
                    return self.socketFrontEndRealtime;
                },
                //#region @backend
                set BE(v) {
                    self.socketNamespaceBE = v;
                },
                get BE() {
                    return self.socketNamespaceBE;
                },
                set BE_REALTIME(v) {
                    self.socketNamespaceBERealtime = v;
                },
                get BE_REALTIME() {
                    return self.socketNamespaceBERealtime;
                }
                //#endregion
            };
        },
        enumerable: true,
        configurable: true
    });
    RealtimeBase.prototype.pathFor = function (namespace) {
        var uri = this.context.uri;
        var nsp = namespace ? namespace : '';
        var pathname = uri.pathname !== '/' ? uri.pathname : '';
        var morphiPrefix = "socketnodejs";
        var href = "" + uri.origin + pathname + "/" + morphiPrefix + nsp;
        // console.log(`HREF: ${href}`)
        return new URL(href);
    };
    return RealtimeBase;
}());
exports.RealtimeBase = RealtimeBase;
