// navi-cli/generator/gen-ts-directive.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenTsDirective = void 0;
var tslib_1 = require("tslib");
var base_generators_backend_1 = require("./base-generators.backend");
var _ = require("lodash");
var GenTsDirective = /** @class */ (function (_super) {
    tslib_1.__extends(GenTsDirective, _super);
    function GenTsDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GenTsDirective.prototype.init = function () {
        _super.prototype.init.call(this);
        if (this.isDecorator) {
            this.className = this.className + "DirectiveDecorator";
        }
        else {
            this.className = this.className + "Directive";
        }
        this.directiveSelector = "" + _.camelCase(this.contextFolderName);
        this.outputFileName = this.outputFileName + ".directive";
    };
    Object.defineProperty(GenTsDirective.prototype, "template", {
        get: function () {
            return "\n" + this.decoratorImport + "\n\nexport function " + this.className + "() {\n    return {\n      restrict: 'A',\n      link: function link($scope, $element, $attrs) {\n\n        try {\n          const config  = $scope.$eval($attrs." + this.directiveSelector + ")\n        } catch(e) {\n\n        }\n\n      }\n    };\n  }\n";
        },
        enumerable: false,
        configurable: true
    });
    return GenTsDirective;
}(base_generators_backend_1.FilesGenerator));
exports.GenTsDirective = GenTsDirective;
