// navi-cli/generator/gen-ts-routing-module.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenTsRoutingModule = void 0;
var tslib_1 = require("tslib");
var _ = require("lodash");
var gen_ts_module_backend_1 = require("./gen-ts-module.backend");
var GenTsRoutingModule = /** @class */ (function (_super) {
    tslib_1.__extends(GenTsRoutingModule, _super);
    function GenTsRoutingModule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(GenTsRoutingModule.prototype, "decoratorImport", {
        get: function () {
            return this.isDecorator ? "import { " + this.className.replace('Module', 'RoutingModule') + " as Base } from '" + this.decoratorModuleFilePath.replace('.decorator', '') + "'" : '';
        },
        enumerable: false,
        configurable: true
    });
    GenTsRoutingModule.prototype.init = function () {
        _super.prototype.init.call(this);
        if (this.isDecorator) {
        }
        else {
            this.className = this.className.replace('Module', 'RoutingModule');
            this.outputFileName = this.outputFileName.replace('.module', '-routing.module');
        }
        this.defaultComponentName = _.lowerFirst(this.className.replace('RoutingModule', 'Component'));
        this.breadcrumb = _.startCase(this.className.replace('RoutingModule', ''));
        var s = _.kebabCase(this.className).split('-');
        var prefix = s.shift();
        s.pop();
        s.pop();
        this.routingUrl = s.join('-');
        this.routingName = prefix + "." + this.routingUrl;
    };
    Object.defineProperty(GenTsRoutingModule.prototype, "template", {
        get: function () {
            return "\nimport { NgModule } from 'es-common/src/decorators';\nimport { Router } from 'es-common/src/components/es-router/es-router-provider';\nimport { EsRouterStateDeclaration } from 'es-common/src/components/es-router/es-router-state-declaration-interface';\n" + this.decoratorImport + "\n\nconst baseState: EsRouterStateDeclaration = {\n  name: '" + this.routingName + "',\n  url: '/" + this.routingUrl + "',\n  abstract: true,\n  template: '<div ui-view></div>',\n  ncyBreadcrumb: {\n    label: gettext('" + this.breadcrumb + "')\n  }\n};\n\nconst viewState: EsRouterStateDeclaration = {\n  name: '" + this.routingName + ".view',\n  url: '/view',\n  component: '" + this.defaultComponentName + "',\n  ncyBreadcrumb: {\n    label: gettext('" + this.breadcrumb + " View'),\n    parent: '" + this.routingName + "'\n  },\n};\n\n@NgModule({\n  id: '" + this.className + "',\n  imports: [\n    Router.forChild([\n      baseState,\n      viewState\n    ]),\n    " + (this.isDecorator ? 'Base,' : '') + "\n  ]\n})\nexport class " + this.className + " {\n}\n\n";
        },
        enumerable: false,
        configurable: true
    });
    return GenTsRoutingModule;
}(gen_ts_module_backend_1.GenTsModule));
exports.GenTsRoutingModule = GenTsRoutingModule;
