// tnp/apps/project/project-ui/tnp-project/items/item-test/item-test.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
//#region angular
var core_1 = require("@angular/core");
//#endregion
//#region isomorphic
var base_item_stepper_1 = require("../base-item-stepper");
var tabs_menu_tnp_project_1 = require("../../tabs-menu-tnp-project");
//#endregion
var ItemTestComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ItemTestComponent, _super);
    function ItemTestComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ItemTestComponent.prototype.tabSelectedAction = function () {
    };
    ItemTestComponent.prototype.formValueChanged = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    ItemTestComponent.prototype.tabNumber = function () {
        return tabs_menu_tnp_project_1.TnpProjectTabIndex.TEST;
    };
    ItemTestComponent.prototype.ngOnInit = function () {
    };
    ItemTestComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-item-test',
            templateUrl: './item-test.component.html',
            styleUrls: ['./item-test.component.scss']
        })
    ], ItemTestComponent);
    return ItemTestComponent;
}(base_item_stepper_1.BaseItemStepperProcessBuildComponent));
exports.ItemTestComponent = ItemTestComponent;
