// tnp/scripts/COMPILERS-TESTING/COMPILERS.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var project_1 = require("../../project");
var tnp_helpers_1 = require("tnp-helpers");
var testTasks = 'Test task';
function $COMPILER_BSJ() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.join.start(testTasks)];
                case 1:
                    _a.sent();
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_BSJ_WATCH() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.join.startAndWatch(testTasks)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_BSJ_WATCH_ONLY() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.join.startAndWatch(testTasks, { watchOnly: true })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_SM() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.sourceModifier.start(testTasks)];
                case 1:
                    _a.sent();
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_SM_WATCH() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.sourceModifier.startAndWatch(testTasks)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_SM_WATCH_ONLY() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.sourceModifier.startAndWatch(testTasks, { watchOnly: true })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_FFG() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    console.info('FRAMEWORK FILES FGENERATOR');
                    return [4 /*yield*/, project_1.Project.Current.frameworkFileGenerator.start(testTasks)];
                case 1:
                    _a.sent();
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_FFG_WATCH() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.frameworkFileGenerator.startAndWatch(testTasks)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function $COMPILER_FFG_WATCH_ONLY() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, project_1.Project.Current.frameworkFileGenerator.startAndWatch(testTasks, { watchOnly: true })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.default = {
    $COMPILER_FFG: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_FFG, '$COMPILER_FFG'),
    $COMPILER_FFG_WATCH: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_FFG_WATCH, '$COMPILER_FFG_WATCH'),
    $COMPILER_FFG_WATCH_ONLY: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_FFG_WATCH_ONLY, '$COMPILER_FFG_WATCH_ONLY'),
    $COMPILER_BSJ: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_BSJ, '$COMPILER_BSJ'),
    $COMPILER_BSJ_WATCH: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_BSJ_WATCH, '$COMPILER_BSJ_WATCH'),
    $COMPILER_BSJ_WATCH_ONLY: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_BSJ_WATCH_ONLY, '$COMPILER_BSJ_WATCH_ONLY'),
    $COMPILER_SM: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_SM, '$COMPILER_SM'),
    $COMPILER_SM_WATCH: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_SM_WATCH, '$COMPILER_SM_WATCH'),
    $COMPILER_SM_WATCH_ONLY: tnp_helpers_1.Helpers.CLIWRAP($COMPILER_SM_WATCH_ONLY, '$COMPILER_SM_WATCH_ONLY'),
};
