// tnp/apps/project/project-ui/tnp-project/tnp-project.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
//#region angular
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var tabs_menu_tnp_project_1 = require("./tabs-menu-tnp-project");
var PROJECT_1 = require("../../PROJECT");
var tnp_helpers_1 = require("tnp-helpers");
//#endregion
var TnpProjectComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TnpProjectComponent, _super);
    function TnpProjectComponent(formBuilder) {
        var _this = _super.call(this) || this;
        _this.formBuilder = formBuilder;
        _this.TnpProjectTabIndex = tabs_menu_tnp_project_1.TnpProjectTabIndex;
        _this.showChildren = false;
        _this.opened = new core_1.EventEmitter();
        _this.closed = new core_1.EventEmitter();
        _this.expanded = false;
        _this.isChild = false;
        return _this;
    }
    TnpProjectComponent.prototype.closeChild = function (name) {
        if (this.expandedChild === name) {
            this.expandedChild = '';
        }
    };
    TnpProjectComponent.prototype.toogleChildren = function () {
        this.showChildren = !this.showChildren;
    };
    Object.defineProperty(TnpProjectComponent.prototype, "keyForProject", {
        get: function () {
            return "show-children-" + this.model.location;
        },
        enumerable: true,
        configurable: true
    });
    TnpProjectComponent.prototype.ngClass = function (index) {
        if (this.model.isWorkspace) {
            return {
                'bold': (this.model.selectedIndex === index)
            };
        }
        return {
            'bold': (this.model.selectedIndex === index - 1)
        };
    };
    TnpProjectComponent.prototype.selectionChange = function (e) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.model.selectedIndex = e.selectedIndex;
                        if (!(this.model.selectedIndex === tabs_menu_tnp_project_1.TnpProjectTabIndex.ENV)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.model.updateEndGetEnvironments()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    TnpProjectComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // if (this.model && this.model.isWorkspace) {
                        this.environmentFormGroup = this.formBuilder.group({
                            hidden: ['', forms_1.Validators.required]
                        });
                        // }
                        this.buildFormGroup = this.formBuilder.group({
                        // hidden: ['', Validators.required]
                        });
                        this.testFormGroup = this.formBuilder.group({
                        // hidden: ['', Validators.required]
                        });
                        this.serveFormGroup = this.formBuilder.group({
                        // hidden: ['', Validators.required]
                        });
                        this.releaseFormGroup = this.formBuilder.group({
                        // hidden: ['', Validators.required]
                        });
                        if (!this.model.isWorkspace) {
                            this.showChildren = true;
                        }
                        if (!this.model.isWorkspace) return [3 /*break*/, 2];
                        this.expanded = !!window.localStorage.getItem(this.keyForProject);
                        if (!this.expanded) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.updateModel()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        this.handlers.push(this.opened.subscribe(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        window.localStorage.setItem(this.keyForProject, 'true');
                                        if (!this.expanded) return [3 /*break*/, 2];
                                        return [4 /*yield*/, this.updateModel()];
                                    case 1:
                                        _a.sent();
                                        _a.label = 2;
                                    case 2: return [2 /*return*/];
                                }
                            });
                        }); }));
                        this.handlers.push(this.closed.subscribe(function () {
                            window.localStorage.removeItem(_this.keyForProject);
                        }));
                        return [2 /*return*/];
                }
            });
        });
    };
    TnpProjectComponent.prototype.updateModel = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.model.updaetAndGetProceses()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.model.updateEndGetEnvironments()];
                    case 2:
                        _a.sent();
                        if (this.model.selectedEnv && this.model.selectedEnv.length > 2) {
                            this.model.selectedIndex = tabs_menu_tnp_project_1.TnpProjectTabIndex.ENV;
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", Object)
    ], TnpProjectComponent.prototype, "opened", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", Object)
    ], TnpProjectComponent.prototype, "closed", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], TnpProjectComponent.prototype, "expanded", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], TnpProjectComponent.prototype, "isChild", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", PROJECT_1.PROJECT)
    ], TnpProjectComponent.prototype, "model", void 0);
    TnpProjectComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-tnp-project',
            templateUrl: './tnp-project.component.html',
            styleUrls: ['./tnp-project.component.scss']
        }),
        tslib_1.__metadata("design:paramtypes", [forms_1.FormBuilder])
    ], TnpProjectComponent);
    return TnpProjectComponent;
}(tnp_helpers_1.BaseComponent));
exports.TnpProjectComponent = TnpProjectComponent;
