// isomorphic-storage/old/service/webstorage.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var webstorage_utility_1 = require("../utility/webstorage.utility");
var WebStorageService = /** @class */ (function () {
    function WebStorageService(storage) {
        this.storage = storage;
    }
    WebStorageService.prototype.get = function (key) {
        return webstorage_utility_1.WebStorageUtility.get(this.storage, key);
    };
    WebStorageService.prototype.set = function (key, value) {
        webstorage_utility_1.WebStorageUtility.set(this.storage, key, value);
    };
    WebStorageService.prototype.remove = function (key) {
        webstorage_utility_1.WebStorageUtility.remove(this.storage, key);
    };
    WebStorageService.prototype.clear = function () {
        for (var _i = 0, _a = this.constructor.keys; _i < _a.length; _i++) {
            var key = _a[_i];
            this.storage.removeItem(key);
        }
    };
    return WebStorageService;
}());
exports.WebStorageService = WebStorageService;
var LocalStorageService = /** @class */ (function (_super) {
    tslib_1.__extends(LocalStorageService, _super);
    function LocalStorageService() {
        return _super.call(this, localStorage) || this;
    }
    LocalStorageService.keys = [];
    LocalStorageService = tslib_1.__decorate([
        core_1.Injectable(),
        tslib_1.__metadata("design:paramtypes", [])
    ], LocalStorageService);
    return LocalStorageService;
}(WebStorageService));
exports.LocalStorageService = LocalStorageService;
var SessionStorageService = /** @class */ (function (_super) {
    tslib_1.__extends(SessionStorageService, _super);
    function SessionStorageService() {
        return _super.call(this, sessionStorage) || this;
    }
    SessionStorageService.keys = [];
    SessionStorageService = tslib_1.__decorate([
        core_1.Injectable(),
        tslib_1.__metadata("design:paramtypes", [])
    ], SessionStorageService);
    return SessionStorageService;
}(WebStorageService));
exports.SessionStorageService = SessionStorageService;
