// tnp-ui/formly-buttion-with-action/formly-buttion-with-action.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _ = require("lodash");
var core_1 = require("@angular/core");
var tnp_helpers_1 = require("tnp-helpers");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var FormlyButtionWithActionComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyButtionWithActionComponent, _super);
    function FormlyButtionWithActionComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FormlyButtionWithActionComponent.prototype.action = function () {
        if (_.isFunction(this.field.templateOptions.action)) {
            this.field.templateOptions.action();
        }
    };
    FormlyButtionWithActionComponent.prototype.ngOnInit = function () {
        if (!this.field.templateOptions.icon) {
            this.field.templateOptions.icon = 'home';
        }
        if (!this.field.templateOptions.lable) {
            this.field.templateOptions.lable = 'Button';
        }
    };
    FormlyButtionWithActionComponent = tslib_1.__decorate([
        typescript_class_helpers_1.CLASS.NAME('FormlyButtionWithActionComponent'),
        core_1.Component({
            selector: 'app-formly-buttion-with-action',
            templateUrl: './formly-buttion-with-action.component.html',
            styleUrls: ['./formly-buttion-with-action.component.scss']
        })
    ], FormlyButtionWithActionComponent);
    return FormlyButtionWithActionComponent;
}(tnp_helpers_1.BaseFormlyComponent));
exports.FormlyButtionWithActionComponent = FormlyButtionWithActionComponent;
