// tnp-ui/formly-buttion-with-action/formly-buttion-with-action.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var typescript_class_helpers_1 = require("typescript-class-helpers");
// angular
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
// formly
var core_2 = require("@ngx-formly/core");
// local
var formly_buttion_with_action_component_1 = require("./formly-buttion-with-action.component");
exports.FormlyModuleMod = core_2.FormlyModule.forRoot({
    types: [
        { name: 'button', component: formly_buttion_with_action_component_1.FormlyButtionWithActionComponent },
        { name: typescript_class_helpers_1.CLASS.getName(formly_buttion_with_action_component_1.FormlyButtionWithActionComponent), component: formly_buttion_with_action_component_1.FormlyButtionWithActionComponent }
    ]
});
var angularModules = [
    common_1.CommonModule,
    forms_1.ReactiveFormsModule
];
var FormlyButtionWithActionModule = /** @class */ (function () {
    function FormlyButtionWithActionModule() {
    }
    FormlyButtionWithActionModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays(angularModules, [
                exports.FormlyModuleMod,
            ]),
            exports: [formly_buttion_with_action_component_1.FormlyButtionWithActionComponent],
            declarations: [formly_buttion_with_action_component_1.FormlyButtionWithActionComponent]
        })
    ], FormlyButtionWithActionModule);
    return FormlyButtionWithActionModule;
}());
exports.FormlyButtionWithActionModule = FormlyButtionWithActionModule;
