// tnp-ui/select/select-wrapper.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
// material
var select_1 = require("@angular/material/select");
var list_1 = require("@angular/material/list");
var icon_1 = require("@angular/material/icon");
var progress_spinner_1 = require("@angular/material/progress-spinner");
// other
var ngx_datatable_1 = require("@swimlane/ngx-datatable");
// local
var select_wrapper_component_1 = require("./select-wrapper.component");
// formly
var core_2 = require("@ngx-formly/core");
// import { FormlyMaterialModule, } from '@ngx-formly/material';
var toggle_1 = require("@ngx-formly/material/toggle");
var datepicker_1 = require("@ngx-formly/material/datepicker");
var core_3 = require("@angular/material/core");
var slider_1 = require("@ngx-formly/material/slider");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var angularModules = [
    forms_1.ReactiveFormsModule
];
var materialModules = [
    select_1.MatSelectModule,
    list_1.MatListModule,
    icon_1.MatIconModule,
    progress_spinner_1.MatProgressSpinnerModule
];
exports.NgxDatatableModuleMod = ngx_datatable_1.NgxDatatableModule.forRoot({
    messages: {
        emptyMessage: 'No data to display',
        totalMessage: 'total',
        selectedMessage: 'selected' // Footer selected message
    }
});
exports.FormlyModuleMod = core_2.FormlyModule.forRoot({
    types: [
        { name: 'selectwrapper', component: select_wrapper_component_1.SelectWrapperComponent },
        { name: typescript_class_helpers_1.CLASS.getName(select_wrapper_component_1.SelectWrapperComponent), component: select_wrapper_component_1.SelectWrapperComponent }
    ],
    validationMessages: [
        { name: 'required', message: 'This field is required' },
    ],
});
var moduleOther = [
    exports.NgxDatatableModuleMod
];
var formlyModules = [
    // FormlyMaterialModule,
    exports.FormlyModuleMod,
    toggle_1.FormlyMatToggleModule,
    datepicker_1.FormlyMatDatepickerModule,
    core_3.MatNativeDateModule,
    slider_1.FormlyMatSliderModule
];
var SelectWrapperModule = /** @class */ (function () {
    function SelectWrapperModule() {
    }
    SelectWrapperModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays([
                common_1.CommonModule
            ], angularModules, moduleOther, materialModules, formlyModules),
            exports: [
                select_wrapper_component_1.SelectWrapperComponent
            ],
            declarations: [select_wrapper_component_1.SelectWrapperComponent]
        })
    ], SelectWrapperModule);
    return SelectWrapperModule;
}());
exports.SelectWrapperModule = SelectWrapperModule;
