// tnp-ui/select/select-wrapper.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// angular
var core_1 = require("@angular/core");
var _ = require("lodash");
var morphi_1 = require("morphi");
var ng2_logger_1 = require("ng2-logger");
var tnp_helpers_1 = require("tnp-helpers");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var forms_1 = require("@angular/forms");
exports.CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms_1.NG_VALUE_ACCESSOR,
    useExisting: core_1.forwardRef(function () { return SelectWrapperComponent; }),
    multi: true
};
var log = ng2_logger_1.Log.create('select wrapper');
var SelectWrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SelectWrapperComponent, _super);
    function SelectWrapperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isLoading = false;
        _this.selectOptions = [];
        _this.data = [
            { value: 'dupa', label: 'Onet' },
            { value: 2, label: 'Google' }
        ];
        _this.label = 'default label';
        _this.valueProp = 'id';
        _this.nameProp = 'name';
        return _this;
    }
    SelectWrapperComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rows, error_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _super.prototype.ngOnInit.call(this);
                        if (!this.crud && _.isFunction(_.get(this.field, 'templateOptions.crud'))) {
                            this.crud = this.field.templateOptions.crud;
                        }
                        if (!_.isFunction(this.crud)) return [3 /*break*/, 5];
                        this.isLoading = true;
                        log.i('this.crud.entity', typescript_class_helpers_1.CLASS.describeProperites(this.crud.entity));
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.crud.getAll().received.observable.take(1).toPromise()];
                    case 2:
                        rows = _a.sent();
                        this.initOptions(rows.body.json);
                        this.isLoading = false;
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        this.isLoading = false;
                        return [3 /*break*/, 4];
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        this.initOptions(this.data);
                        _a.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    SelectWrapperComponent.prototype.initOptions = function (rows) {
        var _this = this;
        this.selectOptions = rows.map(function (r) {
            if (!_this.crud) {
                return r;
            }
            return { value: r[_this.valueProp], label: r[_this.nameProp] };
        });
    };
    SelectWrapperComponent.prototype.ngAfterViewInit = function () {
        // log.i('this', this.field);
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", morphi_1.Morphi.CRUD.Base)
    ], SelectWrapperComponent.prototype, "crud", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], SelectWrapperComponent.prototype, "selectOptions", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], SelectWrapperComponent.prototype, "data", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], SelectWrapperComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], SelectWrapperComponent.prototype, "valueProp", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], SelectWrapperComponent.prototype, "nameProp", void 0);
    SelectWrapperComponent = tslib_1.__decorate([
        typescript_class_helpers_1.CLASS.NAME('SelectWrapperComponent'),
        core_1.Component({
            selector: 'app-select-wrapper',
            templateUrl: './select-wrapper.component.html',
            styleUrls: ['./select-wrapper.component.scss'],
            providers: [exports.CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR]
        })
    ], SelectWrapperComponent);
    return SelectWrapperComponent;
}(tnp_helpers_1.BaseFormlyComponent));
exports.SelectWrapperComponent = SelectWrapperComponent;
