// tnp/apps/project/project-ui/project-ui.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
//#region angular
var core_1 = require("@angular/core");
//#endregion
//#region isomorphic
var morphi_1 = require("morphi");
var PROJECT_1 = require("../PROJECT");
var tnp_helpers_1 = require("tnp-helpers");
//#endregion
var DualComponentControllerExtended = /** @class */ (function (_super) {
    tslib_1.__extends(DualComponentControllerExtended, _super);
    function DualComponentControllerExtended() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DualComponentControllerExtended;
}(tnp_helpers_1.DualComponentController));
exports.DualComponentControllerExtended = DualComponentControllerExtended;
var BuildTnpProcessComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BuildTnpProcessComponent, _super);
    function BuildTnpProcessComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.DualComponentController = DualComponentControllerExtended;
        _this.isLinear = false;
        return _this;
    }
    BuildTnpProcessComponent.prototype.ngOnInit = function () {
        // this.firstFormGroup = this._formBuilder.group({
        //   firstCtrl: ['', Validators.required]
        // });
        // this.secondFormGroup = this._formBuilder.group({
        //   secondCtrl: ['', Validators.required]
        // });
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", PROJECT_1.PROJECT)
    ], BuildTnpProcessComponent.prototype, "model", void 0);
    BuildTnpProcessComponent = tslib_1.__decorate([
        morphi_1.Morphi.Formly.RegisterComponentForEntity(PROJECT_1.PROJECT),
        core_1.Component({
            selector: 'app-project-ui',
            templateUrl: './project-ui.component.html',
            styleUrls: ['./project-ui.component.scss']
        })
    ], BuildTnpProcessComponent);
    return BuildTnpProcessComponent;
}(tnp_helpers_1.BaseFormlyComponent));
exports.BuildTnpProcessComponent = BuildTnpProcessComponent;
