// tnp/apps/project/project-ui/project-ui.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
// other
var typescript_class_helpers_1 = require("typescript-class-helpers");
// material
var button_1 = require("@angular/material/button");
var input_1 = require("@angular/material/input");
var icon_1 = require("@angular/material/icon");
var slide_toggle_1 = require("@angular/material/slide-toggle");
var dialog_1 = require("@angular/material/dialog");
var progress_bar_1 = require("@angular/material/progress-bar");
var stepper_1 = require("@angular/material/stepper");
var form_field_1 = require("@angular/material/form-field");
var expansion_1 = require("@angular/material/expansion");
var card_1 = require("@angular/material/card");
var radio_1 = require("@angular/material/radio");
// local
var project_ui_component_1 = require("./project-ui.component");
var tnp_project_component_1 = require("./tnp-project/tnp-project.component");
var items_1 = require("./tnp-project/items");
var core_2 = require("@ngx-formly/core");
var process_logger_1 = require("../../process/process-logger");
var tnp_ui_1 = require("tnp-ui");
var angularModules = [
    common_1.CommonModule,
    forms_1.ReactiveFormsModule,
];
var materialModules = [
    button_1.MatButtonModule,
    input_1.MatInputModule,
    icon_1.MatIconModule,
    dialog_1.MatDialogModule,
    slide_toggle_1.MatSlideToggleModule,
    progress_bar_1.MatProgressBarModule,
    stepper_1.MatStepperModule,
    form_field_1.MatFormFieldModule,
    expansion_1.MatExpansionModule,
    card_1.MatCardModule,
    radio_1.MatRadioModule
];
var stepperItemsComponents = [
    items_1.ItemEnvironmentComponent,
    items_1.ItemBuildComponent,
    items_1.ItemServeComponent,
    items_1.ItemTestComponent
];
var componentsLocal = tslib_1.__spreadArrays(stepperItemsComponents, [
    tnp_project_component_1.TnpProjectComponent,
    project_ui_component_1.BuildTnpProcessComponent,
]);
var baselineModules = [
    process_logger_1.ProcessLoggerModule,
    tnp_ui_1.ListWrapperModule,
];
var BuildTnpProcessModule = /** @class */ (function () {
    function BuildTnpProcessModule() {
    }
    BuildTnpProcessModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays([
                core_2.FormlyModule.forRoot({
                    types: [
                        { name: 'project', component: project_ui_component_1.BuildTnpProcessComponent },
                        { name: typescript_class_helpers_1.CLASS.getName(project_ui_component_1.BuildTnpProcessComponent), component: project_ui_component_1.BuildTnpProcessComponent },
                    ],
                    validationMessages: [
                        { name: 'required', message: 'This field is required' },
                    ],
                })
            ], angularModules, materialModules, baselineModules),
            exports: tslib_1.__spreadArrays(componentsLocal),
            declarations: tslib_1.__spreadArrays(componentsLocal),
            providers: []
        })
    ], BuildTnpProcessModule);
    return BuildTnpProcessModule;
}());
exports.BuildTnpProcessModule = BuildTnpProcessModule;
