// tnp/scripts/PROJECTS-DEVELOPMENT/BRANDING.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tnp_core_1 = require("tnp-core");
var tnp_core_2 = require("tnp-core");
var svg_png_converter_1 = require("svg-png-converter");
var tnp_helpers_1 = require("tnp-helpers");
var project_1 = require("../../project");
var tnp_config_1 = require("tnp-config");
var files_renaming_backend_1 = require("../../project/features/files-recreation/files-renaming.backend");
function $BRANDING(args, exit) {
    if (exit === void 0) { exit = true; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var proj, pathes, content;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    proj = project_1.Project.Current;
                    pathes = {
                        favicondesc: 'tmp-fav-icon-description.json'
                    };
                    tnp_helpers_1.Helpers.log('Generation svg...');
                    return [4 /*yield*/, svg_png_converter_1.png2svg({
                            tracer: 'imagetracer',
                            optimize: true,
                            input: tnp_core_2.fse.readFileSync(tnp_core_1.path.join(proj.location, tnp_config_1.config.pathes.logoPng)),
                            numberofcolors: 24,
                            pathomit: 1,
                        })];
                case 1:
                    content = (_a.sent()).content;
                    tnp_core_2.fse.writeFileSync(tnp_config_1.config.pathes.logoSvg, content);
                    tnp_helpers_1.Helpers.log('Generation favicons...');
                    tnp_helpers_1.Helpers.writeFile(tnp_core_1.path.join(proj.location, pathes.favicondesc), faviconsDesc());
                    tnp_helpers_1.Helpers.mkdirp(tnp_core_1.path.join(proj.location, 'src/assets/favicons'));
                    proj.run("real-favicon generate " + pathes.favicondesc + " tmp-faviconout.json src/assets/favicons").sync();
                    tnp_helpers_1.Helpers.copyFile(tnp_core_1.path.join(proj.location, 'src/assets/favicons/favicon.ico'), tnp_core_1.path.join(proj.location, 'src/favicon.ico'));
                    tnp_helpers_1.Helpers.log('Generation pwa icons...');
                    proj.run("ngx-pwa-icons --icon ./" + tnp_config_1.config.pathes.logoPng, { output: false }).sync();
                    // proj.run(` pwa-asset-generator ./${config.pathes.logoSvg} -f --icon-only `, { output: false }).sync();
                    tnp_helpers_1.Helpers.info('DONE');
                    if (exit) {
                        process.exit(0);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.$BRANDING = $BRANDING;
function $RENAMING(aa, exit) {
    if (exit === void 0) { exit = true; }
    files_renaming_backend_1.FilesRenaming.start(aa);
    if (exit) {
        process.exit(0);
    }
}
exports.$RENAMING = $RENAMING;
function faviconsDesc() {
    return {
        "masterPicture": tnp_config_1.config.pathes.logoSvg,
        "iconsPath": "/",
        "design": {
            "ios": {
                "pictureAspect": "noChange",
                "assets": {
                    "ios6AndPriorIcons": false,
                    "ios7AndLaterIcons": false,
                    "precomposedIcons": false,
                    "declareOnlyDefaultIcon": true
                }
            },
            "desktopBrowser": {},
            "windows": {
                "pictureAspect": "noChange",
                "backgroundColor": "#da532c",
                "onConflict": "override",
                "assets": {
                    "windows80Ie10Tile": false,
                    "windows10Ie11EdgeTiles": {
                        "small": false,
                        "medium": true,
                        "big": false,
                        "rectangle": false
                    }
                }
            },
            "androidChrome": {
                "pictureAspect": "backgroundAndMargin",
                "margin": "17%",
                "backgroundColor": "#ffffff",
                "themeColor": "#ffffff",
                "manifest": {
                    "display": "standalone",
                    "orientation": "notSet",
                    "onConflict": "override",
                    "declared": true
                },
                "assets": {
                    "legacyIcon": false,
                    "lowResolutionIcons": false
                }
            },
            "safariPinnedTab": {
                "pictureAspect": "silhouette",
                "themeColor": "#5bbad5"
            }
        },
        "settings": {
            "scalingAlgorithm": "Mitchell",
            "errorOnImageTooSmall": false,
            "readmeFile": false,
            "htmlCodeFile": false,
            "usePathAsIs": false
        }
    };
}
exports.faviconsDesc = faviconsDesc;
exports.default = {
    $BRANDING: tnp_helpers_1.Helpers.CLIWRAP($BRANDING, '$BRANDING'),
    $RENAMING: tnp_helpers_1.Helpers.CLIWRAP($RENAMING, '$RENAMING'),
};
