// tnp-ui/list/list-wrapper.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var router_1 = require("@angular/router");
// material
var select_1 = require("@angular/material/select");
var list_1 = require("@angular/material/list");
var icon_1 = require("@angular/material/icon");
var card_1 = require("@angular/material/card");
var button_1 = require("@angular/material/button");
var progress_spinner_1 = require("@angular/material/progress-spinner");
// other
var ngx_datatable_1 = require("@swimlane/ngx-datatable");
// local
var list_wrapper_component_1 = require("./list-wrapper.component");
var core_2 = require("@ngx-formly/core");
var typescript_class_helpers_1 = require("typescript-class-helpers");
var materialModules = [
    select_1.MatSelectModule,
    list_1.MatListModule,
    icon_1.MatIconModule,
    progress_spinner_1.MatProgressSpinnerModule,
    button_1.MatButtonModule,
    card_1.MatCardModule
];
exports.NgxDatatableModuleMod = ngx_datatable_1.NgxDatatableModule.forRoot({
    messages: {
        emptyMessage: 'No data to display',
        totalMessage: 'total',
        selectedMessage: 'selected' // Footer selected message
    }
});
exports.FormlyModuleMod = core_2.FormlyModule.forRoot({
    types: [
        { name: 'listwrapper', component: list_wrapper_component_1.ListWrapperComponent },
        { name: typescript_class_helpers_1.CLASS.getName(list_wrapper_component_1.ListWrapperComponent), component: list_wrapper_component_1.ListWrapperComponent }
    ],
    validationMessages: [
        { name: 'required', message: 'This field is required' },
    ],
});
var moduleOther = [
    exports.NgxDatatableModuleMod
];
var ListWrapperModule = /** @class */ (function () {
    function ListWrapperModule() {
    }
    ListWrapperModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays([
                common_1.CommonModule,
                router_1.RouterModule,
                exports.FormlyModuleMod
            ], moduleOther, materialModules),
            exports: [
                list_wrapper_component_1.ListWrapperComponent
            ],
            declarations: [list_wrapper_component_1.ListWrapperComponent]
        })
    ], ListWrapperModule);
    return ListWrapperModule;
}());
exports.ListWrapperModule = ListWrapperModule;
