// isomorphic-storage/old/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var webstorage_service_1 = require("./service/webstorage.service");
var webstorage_1 = require("./decorator/webstorage");
exports.WebStorage = webstorage_1.WebStorage;
exports.LocalStorage = webstorage_1.LocalStorage;
exports.SessionStorage = webstorage_1.SessionStorage;
var webstorage_service_2 = require("./service/webstorage.service");
exports.WebStorageService = webstorage_service_2.WebStorageService;
exports.LocalStorageService = webstorage_service_2.LocalStorageService;
exports.SessionStorageService = webstorage_service_2.SessionStorageService;
var webstorage_utility_1 = require("./utility/webstorage.utility");
exports.WebStorageUtility = webstorage_utility_1.WebStorageUtility;
tslib_1.__exportStar(require("./config"), exports);
var WebStorageModule = /** @class */ (function () {
    function WebStorageModule() {
    }
    WebStorageModule = tslib_1.__decorate([
        core_1.NgModule({
            providers: [webstorage_service_1.LocalStorageService, webstorage_service_1.SessionStorageService]
        })
    ], WebStorageModule);
    return WebStorageModule;
}());
exports.WebStorageModule = WebStorageModule;
