// lodash-walk-object/tests/bfs-walk-object-by.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var mocha_1 = require("mocha");
var chai_1 = require("chai");
var index_1 = require("../index");
var ng2_logger_1 = require("ng2-logger");
var log = ng2_logger_1.Log.create('bfs walk by');
// const instance = BrowserDB.instance;
mocha_1.describe('BFS Walk.ObjectBy', function () {
    mocha_1.it('BFS Should go through all properties depp in normal object', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var d, expectedPathes, actualPathes, context;
        return tslib_1.__generator(this, function (_a) {
            d = {
                a: {
                    c: 23
                },
                super: 'amazing',
                names: [
                    {
                        name: 'Dariusz'
                    }
                ]
            };
            expectedPathes = [
                '',
                'a',
                'super',
                'names',
                'a.c',
                'names[0]',
                'names[0].name'
            ];
            actualPathes = [];
            context = {
                d: d
            };
            index_1.Helpers.Walk.ObjectBy('d', context, function (v, lodashPath) {
                actualPathes.push(lodashPath);
                // console.log(`Path: "${lodashPath}" `, v)
            }, { breadthWalk: true });
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('BFS Should go through all properties depp in array', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var d, c, arr, expectedPathes, actualPathes, context;
        return tslib_1.__generator(this, function (_a) {
            d = {
                a: {
                    c: 23
                },
                super: 'amazing',
                names: [
                    {
                        name: 'Dariusz'
                    }
                ]
            };
            c = {
                a2: {
                    c2: 25
                },
                super2: 'amazing2',
                names2: [
                    {
                        name: 'Dariusz2'
                    }
                ]
            };
            arr = [
                d,
                c
            ];
            expectedPathes = [
                "",
                "[0]",
                "[1]",
                "[0].a",
                "[0].super",
                "[0].names",
                "[1].a2",
                "[1].super2",
                "[1].names2",
                "[0].a.c",
                "[0].names[0]",
                "[1].a2.c2",
                "[1].names2[0]",
                "[0].names[0].name",
                "[1].names2[0].name"
            ];
            actualPathes = [];
            context = {
                arr: arr
            };
            index_1.Helpers.Walk.ObjectBy('arr', context, function (v, lodashPath) {
                actualPathes.push(lodashPath);
            }, { breadthWalk: true });
            // log.i('actualPathes',actualPathes)
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('BFS Should go through all properties , getters included', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var person, expectedPathes, actualPathes;
        return tslib_1.__generator(this, function (_a) {
            person = {
                name: "Dariusz",
                surname: "Filipiak"
            };
            Object.defineProperty(person, 'fullName', {
                get: function () {
                    return this.name + ' ' + this.surname;
                }
            });
            expectedPathes = ['name', 'surname', 'fullName'];
            actualPathes = [];
            index_1.Helpers.Walk.Object(person, function (v, lodashPath) {
                actualPathes.push(lodashPath);
            }, { walkGetters: true, breadthWalk: true });
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
});
