// lodash-walk-object/tests/dfs-walk-object-by.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var mocha_1 = require("mocha");
var chai_1 = require("chai");
var index_1 = require("../index");
// const instance = BrowserDB.instance;
mocha_1.describe('DFS Walk.ObjectBy', function () {
    mocha_1.it('DFS Should go through all properties depp in normal object', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var d, expectedPathes, actualPathes, context;
        return tslib_1.__generator(this, function (_a) {
            d = {
                a: {
                    c: 23
                },
                super: 'amazing',
                names: [
                    {
                        name: 'Dariusz'
                    }
                ]
            };
            expectedPathes = [
                '',
                'a',
                'a.c',
                'super',
                'names',
                'names[0]',
                'names[0].name'
            ];
            actualPathes = [];
            context = {
                d: d
            };
            index_1.Helpers.Walk.ObjectBy('d', context, function (v, lodashPath) {
                actualPathes.push(lodashPath);
                // console.log(`Path: "${lodashPath}" `, v)
            });
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('DFS Should go through all properties depp in array', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var d, c, arr, expectedPathes, actualPathes, context;
        return tslib_1.__generator(this, function (_a) {
            d = {
                a: {
                    c: 23
                },
                super: 'amazing',
                names: [
                    {
                        name: 'Dariusz'
                    }
                ]
            };
            c = {
                a2: {
                    c2: 25
                },
                super2: 'amazing2',
                names2: [
                    {
                        name: 'Dariusz2'
                    }
                ]
            };
            arr = [
                d,
                c
            ];
            expectedPathes = [
                '',
                '[0]',
                '[0].a',
                '[0].a.c',
                '[0].super',
                '[0].names',
                '[0].names[0]',
                '[0].names[0].name',
                '[1]',
                '[1].a2',
                '[1].a2.c2',
                '[1].super2',
                '[1].names2',
                '[1].names2[0]',
                '[1].names2[0].name'
            ];
            actualPathes = [];
            context = {
                arr: arr
            };
            index_1.Helpers.Walk.ObjectBy('arr', context, function (v, lodashPath) {
                actualPathes.push(lodashPath);
            });
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
    mocha_1.it('DFS Should go through all properties , getters included', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var person, expectedPathes, actualPathes;
        return tslib_1.__generator(this, function (_a) {
            person = {
                name: "Dariusz",
                surname: "Filipiak"
            };
            Object.defineProperty(person, 'fullName', {
                get: function () {
                    return this.name + ' ' + this.surname;
                }
            });
            expectedPathes = ['name', 'surname', 'fullName'];
            actualPathes = [];
            index_1.Helpers.Walk.Object(person, function (v, lodashPath) {
                actualPathes.push(lodashPath);
            }, { walkGetters: true });
            chai_1.expect(expectedPathes).to.deep.eq(actualPathes);
            return [2 /*return*/];
        });
    }); });
});
