// tnp-ui/table-wrapper/table-wrapper.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
//material
var select_1 = require("@angular/material/select");
// other
var ngx_datatable_1 = require("@swimlane/ngx-datatable");
// local
var table_wrapper_component_1 = require("./table-wrapper.component");
var materialModules = [
    select_1.MatSelectModule
];
exports.NgxDatatableModuleMod = ngx_datatable_1.NgxDatatableModule.forRoot({
    messages: {
        emptyMessage: 'No data to display',
        totalMessage: 'total',
        selectedMessage: 'selected' // Footer selected message
    }
});
var moduleOther = [
    exports.NgxDatatableModuleMod
];
var TableWrapperModule = /** @class */ (function () {
    function TableWrapperModule() {
    }
    TableWrapperModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays([
                common_1.CommonModule
            ], moduleOther, materialModules),
            exports: [
                ngx_datatable_1.NgxDatatableModule,
                table_wrapper_component_1.TableWrapperComponent
            ],
            declarations: [table_wrapper_component_1.TableWrapperComponent]
        })
    ], TableWrapperModule);
    return TableWrapperModule;
}());
exports.TableWrapperModule = TableWrapperModule;
