// morphi/build-tool/new.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//#region @backend
var tnp_core_1 = require("tnp-core");
var tnp_cli_1 = require("tnp-cli");
function copyExampleTo(folder, exampleType) {
    var options = {
        overwrite: true,
        recursive: true,
        errorOnExist: true,
        filter: function (src) {
            return !/.*node_modules.*/g.test(src);
        }
    };
    var example = {
        dist: tnp_core_1.crossPlatformPath(tnp_core_1.path.join(tnp_core_1.crossPlatformPath(__dirname), '..', '..', exampleType)),
        bundle: tnp_core_1.crossPlatformPath(tnp_core_1.path.join(tnp_core_1.crossPlatformPath(__dirname), '..', exampleType))
    };
    var destinationPath = tnp_core_1.crossPlatformPath(tnp_core_1.path.join(tnp_core_1.crossPlatformPath(process.cwd()), folder));
    var distMode = tnp_core_1.fse.existsSync(example.dist);
    var sourcePath = distMode ? example.dist : example.bundle;
    sourcePath = tnp_core_1.crossPlatformPath(sourcePath);
    destinationPath = tnp_core_1.crossPlatformPath(destinationPath);
    console.log(tnp_cli_1.CLI.chalk.green("Creating example structure... please wait."));
    if (distMode) {
        tnp_core_1.fse.copySync(sourcePath, destinationPath, options);
    }
    else {
        tnp_core_1.child_process.execSync("cpr \"" + sourcePath + "\" \"" + destinationPath + "\" --owerwrite");
    }
    // console.log(chalk.green(`Morphi example structure created sucessfully, installing npm...`));
    // child.execSync('npm i', { cwd: destinationPath })
    console.log(tnp_cli_1.CLI.chalk.green('Done.'));
    if (tnp_cli_1.CLI.commandExistsSync('code')) {
        tnp_core_1.child_process.execSync('code .', { cwd: destinationPath });
    }
}
exports.copyExampleTo = copyExampleTo;
//#endregion
