// ng-talkback/features/error-rate.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var options_backend_1 = require("../options.backend");
var ErrorRate = /** @class */ (function () {
    function ErrorRate(options) {
        this.options = options;
    }
    ErrorRate.prototype.shouldSimulate = function (req, tape) {
        var globalErrorRate = typeof (this.options.errorRate) === 'number' ? this.options.errorRate : this.options.errorRate(req);
        var errorRate = tape && tape.meta.errorRate !== undefined ? tape.meta.errorRate : globalErrorRate;
        options_backend_1.default.validateErrorRate(errorRate);
        var random = Math.random() * 100;
        return random < errorRate;
    };
    ErrorRate.prototype.simulate = function (req) {
        this.options.logger.log("Simulating error for " + req.url);
        return {
            status: 503,
            headers: { 'content-type': ['text/plain'] },
            body: Buffer.from("talkback - failure injection")
        };
    };
    return ErrorRate;
}());
exports.default = ErrorRate;
