// tnp/apps/project/project-ui/tnp-project/items/item-build/item-build.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tnp_core_1 = require("tnp-core");
var core_1 = require("@angular/core");
var base_item_stepper_1 = require("../base-item-stepper");
var PROJECT_1 = require("../../../../PROJECT");
var tabs_menu_tnp_project_1 = require("../../tabs-menu-tnp-project");
var ItemBuildComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ItemBuildComponent, _super);
    function ItemBuildComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ItemBuildComponent.prototype.tabNumber = function () {
        return tabs_menu_tnp_project_1.TnpProjectTabIndex.BUILD;
    };
    ItemBuildComponent.prototype.formValueChanged = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    ItemBuildComponent.prototype.tabSelectedAction = function () {
        var proc = this.model.procStaticBuild;
        var parameters = tnp_core_1._.isString(this.model.selectedEnv) ? { env: this.model.selectedEnv } : void 0;
        this.model.procStaticBuild.parameters = parameters;
        // const COMMAND_TO_EXECUTE = proc.parameters ? Helpers
        //   .interpolateString(proc.cmd)
        //   .withParameters(proc.parameters)
        //   : proc.cmd;
        // proc.cmd = COMMAND_TO_EXECUTE;
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", PROJECT_1.PROJECT)
    ], ItemBuildComponent.prototype, "model", void 0);
    ItemBuildComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-item-build',
            templateUrl: './item-build.component.html',
            styleUrls: ['./item-build.component.scss']
        })
    ], ItemBuildComponent);
    return ItemBuildComponent;
}(base_item_stepper_1.BaseItemStepperProcessBuildComponent));
exports.ItemBuildComponent = ItemBuildComponent;
