// tnp/apps/process/process-logger/process-info-message/process-info-message.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var PROCESS_1 = require("../../PROCESS");
require("rxjs/add/observable/fromEvent");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var process_logger_base_class_1 = require("../../process-logger-base.class");
//#endregion
var ProcessInfoMessageComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ProcessInfoMessageComponent, _super);
    function ProcessInfoMessageComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.messages = [];
        return _this;
    }
    Object.defineProperty(ProcessInfoMessageComponent.prototype, "process", {
        get: function () {
            return this.model;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProcessInfoMessageComponent.prototype, "lsKey", {
        get: function () {
            return "process-info-message-model-height-" + this.model.id;
        },
        enumerable: true,
        configurable: true
    });
    ProcessInfoMessageComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.handlers.push(this.changes.subscribe(function () {
            // console.log('CHANGES PROCESS INFO')
            _this.messages = _this.model.allProgressData;
            _this.scrollDown();
            _this.removeDuplicates();
        }));
        // console.log(`SUBSCRIBE PROPTERY: ${this.process.id} - allProgressData `);
        // this.process.subscribeRealtimeUpdates({
        //   property: 'allProgressData',
        //   bufforProperty: '_allProgressData',
        //   callback: () => {
        //     this.changes.next(void 0);
        //   }
        // });
    };
    ProcessInfoMessageComponent.prototype.removeDuplicates = function () {
        // if (this.process && this.process._allProgressData) {
        //   const unique = {};
        //   this.process._allProgressData.forEach(c => {
        //     unique[_.snakeCase(c.date as any)] = c;
        //   });
        //   this.process._allProgressData = Object
        //     .keys(unique)
        //     .map(k => unique[k]);
        // }
    };
    ProcessInfoMessageComponent.prototype.ngOnDestroy = function () {
        // this.handlers.forEach(h => h.unsubscribe());
        // this.resizeService.removeResizeEventListener(this.elemetRef.nativeElement);
        // if (this.model instanceof PROCESS) {
        //   this.process.unsubscribeRealtimeUpdates('allProgressData');
        // }
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", PROCESS_1.PROCESS)
    ], ProcessInfoMessageComponent.prototype, "model", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", BehaviorSubject_1.BehaviorSubject)
    ], ProcessInfoMessageComponent.prototype, "changes", void 0);
    ProcessInfoMessageComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-process-info-message',
            templateUrl: './process-info-message.component.html',
            styleUrls: ['process-info-message.component.scss'],
        })
    ], ProcessInfoMessageComponent);
    return ProcessInfoMessageComponent;
}(process_logger_base_class_1.ProcessLoggerBaseClass));
exports.ProcessInfoMessageComponent = ProcessInfoMessageComponent;
