// morphi/decorators/decorators-params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var models_1 = require("../models");
var typescript_class_helpers_1 = require("typescript-class-helpers");
function metaParam(param, name, expire, defaultValue) {
    if (defaultValue === void 0) { defaultValue = undefined; }
    var args = [];
    for (var _i = 4; _i < arguments.length; _i++) {
        args[_i - 4] = arguments[_i];
    }
    var methodName = args[1];
    var configs = typescript_class_helpers_1.CLASS.getConfig(args[0].constructor);
    var c = configs[0];
    var m = c.methods[methodName] = (!c.methods[methodName] ? new models_1.Models.Rest.MethodConfig() : c.methods[methodName]);
    var nameKey = name ? name : param;
    var p = m.parameters[nameKey] = (!m.parameters[nameKey] ? new models_1.Models.Rest.ParamConfig() : m.parameters[nameKey]);
    p.index = args[2];
    p.paramName = name;
    p.paramType = param;
    p.defaultType = defaultValue;
    p.expireInSeconds = expire;
}
function Path(name) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        metaParam.apply(void 0, tslib_1.__spreadArrays(['Path', name, undefined, {}], args));
    };
}
exports.Path = Path;
function Query(name) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        metaParam.apply(void 0, tslib_1.__spreadArrays(['Query', name, undefined, {}], args));
    };
}
exports.Query = Query;
function Cookie(name, expireInSecond) {
    if (expireInSecond === void 0) { expireInSecond = 3600; }
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        metaParam.apply(void 0, tslib_1.__spreadArrays(['Cookie', name, expireInSecond, {}], args));
    };
}
exports.Cookie = Cookie;
function Header(name) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        metaParam.apply(void 0, tslib_1.__spreadArrays(['Header', name, undefined, {}], args));
    };
}
exports.Header = Header;
function Body(name) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        metaParam.apply(void 0, tslib_1.__spreadArrays(['Body', name, undefined, {}], args));
    };
}
exports.Body = Body;
