// tnp/apps/process/process-logger/process-logger.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
//#region angular
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
//#endregion
//#region material
var expansion_1 = require("@angular/material/expansion");
var tabs_1 = require("@angular/material/tabs");
var icon_1 = require("@angular/material/icon");
var list_1 = require("@angular/material/list");
var progress_spinner_1 = require("@angular/material/progress-spinner");
//#endregion
//#region local
var process_logger_component_1 = require("./process-logger.component");
var process_console_info_module_1 = require("./process-console-info/process-console-info.module");
var process_info_message_component_1 = require("./process-info-message/process-info-message.component");
//#endregion
//#region isomorphic
var tnp_ui_1 = require("tnp-ui");
var tnp_ui_2 = require("tnp-ui");
var tnp_helpers_1 = require("tnp-helpers");
var typescript_class_helpers_1 = require("typescript-class-helpers");
//#endregion
var ngx_moment_1 = require("ngx-moment");
var core_2 = require("@ngx-formly/core");
var tnp_helpers_2 = require("tnp-helpers");
var angularModules = [
    common_1.CommonModule,
    forms_1.FormsModule,
    forms_1.ReactiveFormsModule,
];
var materialModules = [
    expansion_1.MatExpansionModule,
    tabs_1.MatTabsModule,
    icon_1.MatIconModule,
    list_1.MatListModule,
    progress_spinner_1.MatProgressSpinnerModule
];
var otherModules = [
    ngx_moment_1.MomentModule
];
var localModules = [
    tnp_ui_2.DraggablePopupModule,
    tnp_ui_1.ButtonIconModule,
    process_console_info_module_1.ProcessConsoleInfoModule,
    core_2.FormlyModule.forRoot({
        types: [
            { name: typescript_class_helpers_1.CLASS.getName(process_logger_component_1.ProcessLoggerComponent), component: process_logger_component_1.ProcessLoggerComponent },
        ],
        validationMessages: [
            { name: 'required', message: 'This field is required' },
        ],
    }),
];
var localComponents = [
    process_info_message_component_1.ProcessInfoMessageComponent,
    process_logger_component_1.ProcessLoggerComponent
];
var ProcessLoggerModule = /** @class */ (function () {
    function ProcessLoggerModule() {
    }
    ProcessLoggerModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays(angularModules, materialModules, otherModules, localModules),
            entryComponents: [process_logger_component_1.ProcessLoggerComponent],
            exports: [process_logger_component_1.ProcessLoggerComponent, ngx_moment_1.MomentModule],
            declarations: tslib_1.__spreadArrays([
                tnp_helpers_2.LongPress
            ], localComponents),
            providers: [tnp_helpers_1.ResizeService]
        })
    ], ProcessLoggerModule);
    return ProcessLoggerModule;
}());
exports.ProcessLoggerModule = ProcessLoggerModule;
