// tnp-ui/editor/editor-wrapper.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var ng2_logger_1 = require("ng2-logger");
var log = ng2_logger_1.Log.create('editor wrapper');
var typescript_class_helpers_1 = require("typescript-class-helpers");
var tnp_helpers_1 = require("tnp-helpers");
var DualComponentControllerExtended = /** @class */ (function (_super) {
    tslib_1.__extends(DualComponentControllerExtended, _super);
    function DualComponentControllerExtended() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(DualComponentControllerExtended.prototype, "buttons", {
        get: function () {
            var res = this.getValTemplateOptions('buttons');
            return Array.isArray(res) ? res.toString() : '';
        },
        enumerable: true,
        configurable: true
    });
    return DualComponentControllerExtended;
}(tnp_helpers_1.DualComponentController));
exports.DualComponentControllerExtended = DualComponentControllerExtended;
var EditorWrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(EditorWrapperComponent, _super);
    function EditorWrapperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.DualComponentController = DualComponentControllerExtended;
        _this.buttons = ['bold', 'italic', 'underline'];
        return _this;
    }
    EditorWrapperComponent.prototype.contentChange = function (e) {
        var _this = this;
        setTimeout(function () {
            _this.ctrl.value = e;
            // this.field.formControl.setValue(e)
        });
    };
    EditorWrapperComponent.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], EditorWrapperComponent.prototype, "buttons", void 0);
    EditorWrapperComponent = tslib_1.__decorate([
        typescript_class_helpers_1.CLASS.NAME('EditorWrapperComponent'),
        core_1.Component({
            selector: 'app-editor-wrapper',
            templateUrl: './editor-wrapper.component.html',
            styleUrls: ['./editor-wrapper.component.scss']
        })
    ], EditorWrapperComponent);
    return EditorWrapperComponent;
}(tnp_helpers_1.BaseFormlyComponent));
exports.EditorWrapperComponent = EditorWrapperComponent;
