// tnp-ui/editor/editor-wrapper.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var editor_wrapper_component_1 = require("./editor-wrapper.component");
// formly
var core_2 = require("@ngx-formly/core");
// import { FormlyMaterialModule, } from '@ngx-formly/material';
var toggle_1 = require("@ngx-formly/material/toggle");
var datepicker_1 = require("@ngx-formly/material/datepicker");
var core_3 = require("@angular/material/core");
var slider_1 = require("@ngx-formly/material/slider");
// other
var ngx_wig_1 = require("ngx-wig");
var typescript_class_helpers_1 = require("typescript-class-helpers");
exports.FormlyModuleMod = core_2.FormlyModule.forRoot({
    types: [
        { name: 'texteditor', component: editor_wrapper_component_1.EditorWrapperComponent },
        { name: typescript_class_helpers_1.CLASS.getName(editor_wrapper_component_1.EditorWrapperComponent), component: editor_wrapper_component_1.EditorWrapperComponent }
    ]
});
var formlyModules = [
    // FormlyMaterialModule,
    exports.FormlyModuleMod,
    toggle_1.FormlyMatToggleModule,
    datepicker_1.FormlyMatDatepickerModule,
    core_3.MatNativeDateModule,
    slider_1.FormlyMatSliderModule
];
var otherModules = [
    ngx_wig_1.NgxWigModule
];
var EditorWrapperModule = /** @class */ (function () {
    function EditorWrapperModule() {
    }
    EditorWrapperModule = tslib_1.__decorate([
        core_1.NgModule({
            imports: tslib_1.__spreadArrays([
                common_1.CommonModule
            ], formlyModules, otherModules),
            exports: [
                editor_wrapper_component_1.EditorWrapperComponent
            ],
            declarations: [editor_wrapper_component_1.EditorWrapperComponent]
        })
    ], EditorWrapperModule);
    return EditorWrapperModule;
}());
exports.EditorWrapperModule = EditorWrapperModule;
