// ng2-rest/tests/mapping-test.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var mocha_1 = require("mocha");
var chai_1 = require("chai");
var typescript_class_helpers_1 = require("typescript-class-helpers");
// import { BrowserDB } from '../browser-db/browser-db';
var mapping_1 = require("../mapping");
var ng2_logger_1 = require("ng2-logger");
var log = ng2_logger_1.Log.create('mapping test');
// const instance = BrowserDB.instance;
mocha_1.describe('mapping, default model', function () {
    var testName = 'testname';
    var Test1 = /** @class */ (function () {
        function Test1(age) {
            if (age === void 0) { age = 23; }
            this.age = age;
            this.isHuman = true;
        }
        Test1 = tslib_1.__decorate([
            mapping_1.Mapping.DefaultModelWithMapping({
                name: testName,
                age: 44,
                isHuman: false
            }, {
                parent: 'Test2'
            }),
            tslib_1.__metadata("design:paramtypes", [Number])
        ], Test1);
        return Test1;
    }());
    var Test2 = /** @class */ (function () {
        function Test2() {
        }
        Test2 = tslib_1.__decorate([
            mapping_1.Mapping.DefaultModelWithMapping({}, {
                tests: ['Test1']
            })
        ], Test2);
        return Test2;
    }());
    mocha_1.it('Should generate proper mapping object', function () {
        var mapping = mapping_1.Mapping.getModelsMapping(Test2);
        chai_1.expect(mapping).to.deep.eq({ '': 'Test2', tests: ['Test1'] });
    });
    mocha_1.it('Should generate proper mapping object', function () {
        var mapping = mapping_1.Mapping.getModelsMapping(Test1);
        chai_1.expect(mapping).to.deep.eq({ '': 'Test1', parent: 'Test2' });
    });
});
mocha_1.describe('mapping, nested classes', function () {
    var Company = /** @class */ (function () {
        function Company() {
        }
        Company = tslib_1.__decorate([
            typescript_class_helpers_1.CLASS.NAME('Company'),
            mapping_1.Mapping.DefaultModelWithMapping({ isApple: false })
        ], Company);
        return Company;
    }());
    var Cover = /** @class */ (function () {
        function Cover() {
        }
        Cover = tslib_1.__decorate([
            typescript_class_helpers_1.CLASS.NAME('Cover'),
            mapping_1.Mapping.DefaultModelWithMapping({ thickness: 43 }, { company: 'Company' })
        ], Cover);
        return Cover;
    }());
    var Book = /** @class */ (function () {
        function Book() {
        }
        Book = tslib_1.__decorate([
            typescript_class_helpers_1.CLASS.NAME('Book'),
            mapping_1.Mapping.DefaultModelWithMapping({ title: 'n/a' }, { author: 'User', cover: 'Cover' })
        ], Book);
        return Book;
    }());
    var User = /** @class */ (function () {
        function User() {
        }
        User = tslib_1.__decorate([
            typescript_class_helpers_1.CLASS.NAME('User'),
            mapping_1.Mapping.DefaultModelWithMapping({ name: 'testName' }, { books: ['Book'] })
        ], User);
        return User;
    }());
    mocha_1.it('Should generate proper mapping object', function () {
        var m = mapping_1.Mapping.getModelsMapping(User);
        var res = mapping_1.Mapping.encode({
            name: 'asd',
            books: [
                {
                    title: 'aaaa',
                    cover: {
                        thickness: 77,
                        company: {
                            isApple: true
                        }
                    }
                }
            ]
        }, m);
        console.log('res', res);
        chai_1.expect(res).to.be.instanceOf(User);
        chai_1.expect(res.books[0]).to.be.instanceOf(Book);
        chai_1.expect(res.books[0].cover).to.be.instanceOf(Cover);
        chai_1.expect(res.books[0].cover.company).to.be.instanceOf(Company);
    });
});
