// tnp/project/features/files-recreation/files-templates/base-firedev-file-template.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tnp_core_1 = require("tnp-core");
var BaseFiredevFileTemplate = /** @class */ (function () {
    function BaseFiredevFileTemplate(location, ModuleName) {
        this.location = location;
        this.ModuleName = ModuleName;
        this.FramweorkName = 'Morphi';
    }
    Object.defineProperty(BaseFiredevFileTemplate.prototype, "ControllerName", {
        get: function () {
            return tnp_core_1._.upperFirst(tnp_core_1._.camelCase(this.ModuleName)) + "Controller";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFiredevFileTemplate.prototype, "ENTITYName", {
        get: function () {
            return this.ModuleName.toUpperCase();
        },
        enumerable: true,
        configurable: true
    });
    BaseFiredevFileTemplate.prototype.baseImports = function () {
        return "\nimport { _ } from 'tnp-core';\n//#region @backend\nimport { path } from 'tnp-core'\nimport { fse } from 'tnp-core'\n//#endregion\n    ".trim();
    };
    BaseFiredevFileTemplate.prototype.save = function () {
        if (tnp_core_1._.isFunction(this.files)) {
            var files = this.files();
            if (tnp_core_1._.isArray(files)) {
                for (var index = 0; index < files.length; index++) {
                    var file = files[index];
                    file.save();
                }
            }
        }
    };
    return BaseFiredevFileTemplate;
}());
exports.BaseFiredevFileTemplate = BaseFiredevFileTemplate;
