// navi-cli/current-projects-and-servers.backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllDefaultColors = exports.NavigatorBaselineForksObj = exports.NavigatorBaselineForks = exports.NavigatorBusinessModuleObj = void 0;
var tools_backend_1 = require("./tools.backend");
var NavigatorBaselineForksObject = {
    //#region es-ui
    'es-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:ngin-boa/es-ui.git',
        color: 'blue',
        description: 'Navigator baseline FE ',
        baseline: 'v1.7',
        branch: 'develop',
        codeStyle: 'angular-2+',
        apiProxyTarget: {
            default: {
                'ip': '',
                'domain': 'https://xxa311314'
            },
            other: {
                '10.17.40.94': {
                    'ip': 'https://10.17.40.94',
                    'domain': ''
                },
                'poluan03': {
                    'ip': 'http://172.30.182.109',
                    'domain': ''
                },
                'A3.0 DEV': {
                    'ip': '',
                    'domain': 'http://xxmkt342'
                },
                'xxsce1018': {
                    'ip': '',
                    'domain': 'http://xxsce1018'
                },
                'xxmkt065': {
                    'ip': '',
                    'domain': 'http://xxmkt065'
                },
                'xxmkt149': {
                    'ip': '',
                    'domain': 'http://xxmkt149'
                },
                'A3.1 OLPM product templates': {
                    'ip': '',
                    'domain': 'http://xxmkt202'
                },
                '1.7 DEV': {
                    'ip': '',
                    'domain': 'https://xxa311314'
                },
                '1.7 QA': {
                    'ip': '',
                    'domain': 'https://xxa3010'
                },
                'A3.0': {
                    'ip': '',
                    'domain': 'http://pl-n100iwz'
                },
                'xxmkt1219': {
                    'ip': '',
                    'domain': 'http://xxmkt1219'
                },
                'A3.1 INT': {
                    'ip': '',
                    'domain': 'https://xxmkt448'
                },
                'A3.1 QA': {
                    'ip': '',
                    'domain': 'https://xxmkt1069'
                },
                'xxn18q1a10': {
                    'ip': '',
                    'domain': 'https://xxn18q1a10'
                },
                'may have call statistics data': {
                    'ip': '',
                    'domain': 'https://xxwvs1311'
                },
                'xxmkt1605': {
                    'ip': '',
                    'domain': 'https://xxmkt1605'
                },
                'xxwvs1129.gtk.gtech.com': {
                    'ip': '',
                    'domain': 'https://xxwvs1129.gtk.gtech.com'
                },
                'GA QA': {
                    'ip': 'https://10.17.24.114',
                    'domain': ''
                },
                'sc': {
                    'ip': 'http://10.17.23.78',
                    'domain': ''
                },
                'planogram dev': {
                    'ip': 'http://172.30.182.132',
                    'domain': ''
                },
                'some 1.5/1.4 backend': {
                    'ip': 'http://172.30.182.109',
                    'domain': ''
                }
            }
        },
        addressValidationProxyTarget: {
            default: {
                'ip': 'http://10.17.28.40:8080',
                'domain': ''
            },
            other: {
                '10.17.28.40:8080': {
                    'ip': 'http://10.17.28.40:8080',
                    'domain': ''
                },
                'pl-n100iwz': {
                    'ip': '',
                    'domain': 'http://pl-n100iwz'
                },
                'localhost': {
                    'ip': '',
                    'domain': 'http://localhost'
                }
            },
        },
        schedulerProxyTarget: {
            'other': {
                'xxn18d1arnapp 10.17.45.108': {
                    'ip': 'https://10.17.40.94',
                    'domain': ''
                },
                'A3.1 INT': {
                    'ip': '',
                    'domain': 'https://xxmkt1069'
                }
            },
            'default': {
                'ip': 'https://10.17.40.94',
                'domain': ''
            }
        }
    }),
    //#endregion
    //#region sks-ui
    'sks-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:sks_navigator/sks-ui.git',
        color: 'cornsilk',
        description: 'Slovakia site FE',
        baseline: 'v1.6',
        branch: 'develop',
    }),
    //#endregion
    //#region cas-ui
    'cas-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:cas_navigator/cas-ui.git',
        color: 'goldenrod',
        description: 'California site FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'mixed',
        apiProxyTarget: {
            default: {
                'ip': 'http://10.17.25.6:8180',
                'domain': ''
            },
            other: {
                'Dev 1 App': {
                    'ip': 'http://10.17.25.6:8180',
                    'domain': ''
                },
                'Ips Dev': {
                    'ip': 'http://10.17.18.60:8180',
                    'domain': ''
                },
                'Demo Dev': {
                    'ip': 'http://10.17.37.69:8180',
                    'domain': ''
                },
                'Cap Dev': {
                    'ip': 'http://10.17.32.55:8180',
                    'domain': ''
                },
                'Integ': {
                    'ip': 'http://10.17.25.72:8180',
                    'domain': ''
                },
                'Qa': {
                    'ip': 'http://10.17.16.251:8180',
                    'domain': ''
                },
                'DEV CAP': {
                    'ip': 'http://10.17.37.69:8180',
                    'domain': ''
                },
                'DEV 1': {
                    'ip': '',
                    'domain': 'https://xxa311314'
                },
                'DEV': {
                    'ip': 'http://10.17.25.6:8180',
                    'domain': ''
                },
                'CAP DEV': {
                    'ip': 'http://10.17.32.55:8180',
                    'domain': ''
                },
                'QA': {
                    'ip': '',
                    'domain': 'https://xxcas1147'
                },
                'QA 1': {
                    'ip': '',
                    'domain': 'https://xxcas1112'
                },
                'INTEGRATION': {
                    'ip': '',
                    'domain': 'https://xxcas1185'
                },
                '10 17 18 60 8180': {
                    'ip': 'http://10.17.18.60:8180',
                    'domain': ''
                },
                'Xxcas 107': {
                    'ip': '',
                    'domain': 'https://xxcas107'
                }
            }
        },
        addressValidationProxyTarget: {
            'other': {
                '10 17 28 40 8780': {
                    'ip': 'http://10.17.28.40:8780',
                    'domain': ''
                },
                '10 17 25 6 8780': {
                    'ip': 'http://10.17.25.6:8780',
                    'domain': ''
                },
                '10 17 37 69 8080': {
                    'ip': 'http://10.17.37.69:8080',
                    'domain': ''
                }
            },
            default: {
                'ip': 'http://10.17.25.6:8780',
                'domain': ''
            }
        },
    }),
    //#endregion
    //#region sdl-ui
    'sdl-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:sdl_navigator/sdl-ui.git',
        color: 'darkgreen',
        description: 'South Dakota FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'navigator-angular-js',
    }),
    //#endregion
    //#region swz-ui
    'swz-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:swz_navigator/swz-ui',
        color: 'royalblue',
        description: 'Switzerland FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'angular-2+',
    }),
    //#endregion
    //#region gas-ui
    'gas-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:gas_navigator/gas-ui.git',
        color: 'green',
        description: "Georgia site FE",
        baseline: 'v1.5',
        branch: 'develop',
        codeStyle: 'navigator-angular-js',
        apiProxyTarget: {
            default: {
                'ip': 'https://10.17.24.145:443',
                'domain': ''
            },
            other: {
                'QA': {
                    'ip': 'https://10.17.24.204:443',
                    'domain': ''
                },
                'Dev': {
                    'ip': 'https://10.17.24.145:443',
                    'domain': ''
                }
            }
        },
    }),
    //#endregion
    //#region sce-ui
    'sce-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:sce_navigator/sce-ui.git',
        color: 'violet',
        description: 'South Carolina FE',
        baseline: 'v1.5',
        branch: 'develop',
        codeStyle: 'navigator-angular-js',
        apiProxyTarget: {
            default: {
                'ip': 'https://10.17.22.45:443',
                'domain': ''
            },
            other: {
                'Dev': {
                    'ip': 'https://10.17.22.45:443',
                    'domain': ''
                },
                'Dev 2': {
                    'ip': 'https://10.17.23.78:443',
                    'domain': ''
                },
                'QA': {
                    'ip': 'https://10.17.24.204:443',
                    'domain': ''
                },
                'Int': {
                    'ip': 'https://10.17.24.105:443',
                    'domain': ''
                }
            }
        },
    }),
    //#endregion
    //#region wvs-ui
    'wvs-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:wvs_navigator/wvs-ui.git',
        color: 'lime',
        description: ' West Virginia FE',
        baseline: 'v1.5',
        branch: 'develop',
        codeStyle: 'navigator-angular-js',
        apiProxyTarget: {
            default: {
                'ip': '',
                'domain': 'wvs-dev.localhost:9000'
            },
            other: {
                'Dev': {
                    'ip': '',
                    'domain': 'https://xxwvs1103.gtk.gtech.com:443'
                },
                'QA': {
                    'ip': '',
                    'domain': 'https://xxwvs1129.gtk.gtech.com:443'
                }
            }
        },
    }),
    //#endregion
    //#region vas-ui
    'vas-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:vas_navigator/vas-ui.git',
        color: 'coral',
        description: 'Virginia FE',
        baseline: 'v1.5',
        branch: 'develop',
        codeStyle: 'navigator-angular-js',
        apiProxyTarget: {
            default: {
                domain: 'http://xxvas9.gtk.gtech.com:8180',
                ip: '',
            },
            other: {
                QA: {
                    domain: 'http://xxvas9.gtk.gtech.com:8180',
                    ip: '',
                },
                INT: {
                    domain: 'http://xxvas8.gtk.gtech.com:8180',
                    ip: '',
                },
                DEV: {
                    domain: 'http://xxvas7.gtk.gtech.com:8180',
                    ip: '',
                }
            }
        }
    }),
    //#endregion
    //#region ncl-ui
    'ncl-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:ncl_navigator/ncl-ui',
        color: 'cyan',
        description: 'North Caroline FE',
        baseline: 'v1.5',
        branch: 'batch/cy2003',
        stage: 'production',
        codeStyle: 'navigator-angular-js',
        apiProxyTarget: {
            default: {
                'ip': 'https://10.17.24.145:443',
                'domain': 'sce-dev.localhost:9000'
            },
            other: {
                'QA': {
                    'ip': 'https://10.17.24.204:443',
                    'domain': 'sce-qa.localhost:9000'
                },
                'Dev': {
                    'ip': 'https://10.17.24.145:443',
                    'domain': 'sce-dev.localhost:9000'
                }
            }
        },
    }),
    //#endregion
    //#region czh-ui
    'czh-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:czh_navigator/czh-ui.git',
        color: 'cadetblue',
        description: 'Czech-Republic FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'angular-2+',
        etcHostsAliases: {
            'DEV1 (d1)': {
                ipOrDomain: '10.17.47.89',
                aliases: 'czh-d1-navigator.dev.igt.com czh-d1-bo.dev.igt.com czh-d1-ia.dev.igt.com'
            },
            'DEV2 (d2)': {
                ipOrDomain: '10.17.44.67',
                aliases: 'czh-d2-navigator.dev.igt.com czh-d2-bo.dev.igt.com',
            },
            'DEV3 (d3)': {
                ipOrDomain: '10.17.45.173',
                aliases: 'czh-d3-navigator.dev.igt.com czh-d3-bo.dev.igt.com',
            },
            'DEV5 (d5)': {
                ipOrDomain: '10.17.45.91',
                aliases: 'czh-d5-navigator.dev.igt.com czh-d5-bo.dev.igt.com czh-d5-ia.dev.igt.com',
            },
            'DEV6 (d6)': {
                ipOrDomain: '10.17.44.117',
                aliases: 'czh-d6-navigator.dev.igt.com czh-d6-bo.dev.igt.com czh-d6-ia.dev.igt.com',
            },
            'IPS Dev1 (d7)': {
                ipOrDomain: '10.17.43.246',
                aliases: 'czh-d7-navigator.dev.igt.com czh-d7-bo.dev.igt.com czh-d7-ia.dev.igt.com',
            },
            'INT Dev1 (i1)': {
                ipOrDomain: '10.17.47.14',
                aliases: 'czh-i1-navigator.dev.igt.com czh-i1-bo.dev.igt.com czh-i1-ia.dev.igt.com',
            },
            'QA 1': {
                ipOrDomain: '10.17.41.226',
                aliases: 'czh-q1-navigator.dev.igt.com czh-q1-bo.dev.igt.com czh-q1-ia.dev.igt.com'
            },
            'QA 2': {
                ipOrDomain: '10.17.45.38',
                aliases: 'czh-q2-navigator.dev.igt.com czh-q2-bo.dev.igt.com czh-q2-ia.dev.igt.com'
            },
            'QA 3': {
                ipOrDomain: '10.17.47.62',
                aliases: 'czh-q3-navigator.dev.igt.com czh-q3-bo.dev.igt.com czh-q3-ia.dev.igt.com'
            },
            'QA 4': {
                ipOrDomain: '10.17.39.233',
                aliases: 'czh-q4-navigator.dev.igt.com czh-q4-bo.dev.igt.com czh-q4-ia.dev.igt.com'
            },
            'QA 5': {
                ipOrDomain: '10.17.39.160',
                aliases: 'czh-q5-navigator.dev.igt.com czh-q5-bo.dev.igt.com czh-q5-ia.dev.igt.com'
            },
            'QA 6': {
                ipOrDomain: '10.17.44.236',
                aliases: 'czh-q6-navigator.dev.igt.com czh-q6-bo.dev.igt.com czh-q6-ia.dev.igt.com'
            },
            'QA 7': {
                ipOrDomain: '10.17.32.75',
                aliases: 'czh-q7-navigator.dev.igt.com czh-q7-bo.dev.igt.com czh-q7-ia.dev.igt.com'
            }
        },
        apiProxyTarget: {
            default: {
                domain: 'https://czh-d7-navigator.dev.igt.com',
            },
            other: {
                'DEV1': {
                    domain: 'https://czh-d1-navigator.dev.igt.com'
                },
                'DEV2': {
                    domain: 'https://czh-d2-navigator.dev.igt.com'
                },
                'DEV3': {
                    domain: 'https://czh-d3-navigator.dev.igt.com'
                },
                'INT': {
                    domain: 'https://czh-i1-navigator.dev.igt.com'
                },
                'DEV IPS (D7)': {
                    domain: 'https://czh-d7-navigator.dev.igt.com',
                },
                'QA1': {
                    domain: 'https://czh-q1-navigator.dev.igt.com',
                },
                'QA3': {
                    domain: 'https://czh-q3-navigator.dev.igt.com',
                }
            }
        },
        addressValidationProxyTarget: {
            'other': {
                '10.17.25.6:8780': {
                    'ip': 'http://10.17.25.6:8780',
                    'domain': ''
                }
            },
            'default': {
                'ip': 'http://10.17.25.6:8780',
                'domain': ''
            }
        },
    }),
    //#endregion
    //#region ors-ui
    'ors-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:ors_navigator/ors-ui.git',
        color: 'darkkhaki',
        description: 'Oregon FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'angular-2+',
        apiProxyTarget: {
            default: {
                'ip': '',
                'domain': 'https://ors-d1-navigator.dev.igt.com'
            },
            other: {
                'DEV1': {
                    'ip': '',
                    'domain': 'https://ors-d1-navigator.dev.igt.com'
                },
                'DEV2': {
                    'ip': '',
                    'domain': 'https://ors-d2-navigator.dev.igt.com'
                },
                'CA DEV': {
                    'ip': 'http://10.17.18.60:8180',
                    'domain': ''
                },
                'DEV1/': {
                    'ip': 'http://10.17.38.130:8180',
                    'domain': ''
                },
                'DEV2 IPS/': {
                    'ip': 'http://10.17.38.62:8180',
                    'domain': ''
                },
                'DEV3': {
                    'ip': 'http://10.17.32.124:8180',
                    'domain': ''
                },
                'INT': {
                    'ip': 'http://10.17.38.102:8180',
                    'domain': ''
                },
                'QA1': {
                    'ip': 'http://10.17.36.108:8180',
                    'domain': ''
                },
                'QA2': {
                    'ip': 'http://10.17.36.36:8180',
                    'domain': ''
                }
            }
        },
        addressValidationProxyTarget: {
            'other': {
                'California address validation proxy': {
                    'ip': 'http://10.17.25.6:8780',
                    'domain': ''
                }
            },
            default: {
                'ip': 'http://10.17.25.6:8780',
                'domain': ''
            }
        },
    }),
    //#endregion
    //#region nes-ui
    'nes-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:nes_navigator/nes-ui.git',
        color: 'crimson',
        description: 'Nebraska FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'angular-2+',
        etcHostsAliases: {
            'DEV 1 (d1)': {
                ipOrDomain: '10.17.49.34',
                aliases: [
                    'nes-d1-navigator.dev.igt.com',
                    'nes-d1-bo.dev.igt.com',
                    'nes-d1-pln.dev.igt.com',
                    'nes-d1-ia.dev.igt.com',
                    'nes-d1-ia-bo.dev.igt.com'
                ],
            },
            'I1 integration': {
                ipOrDomain: '10.17.49.95',
                aliases: 'nes-i1-navigator.dev.igt.com nes-i1-bo.dev.igt.com nes-i1-pln.dev.igt.com  nes-i1-ia.dev.igt.com nes-i1-ia-bo.dev.igt.com'
            }
        },
        apiProxyTarget: {
            default: {
                'domain': 'https://nes-d1-navigator.dev.igt.com',
                'ip': ''
            },
            other: {
                'D1': {
                    'domain': 'https://nes-d1-navigator.dev.igt.com',
                    'ip': ''
                }
            }
        }
    }),
    //#endregion
    //#region plt-ui
    'plt-ui': new tools_backend_1.Tools.NavigatorProjectInfo({
        url: 'git@xxgit1.gtech.com:plt_navigator/plt-ui.git',
        color: 'blueviolet',
        description: 'Poland FE',
        baseline: 'v1.6',
        branch: 'develop',
        codeStyle: 'angular-2+',
        etcHostsAliases: {
            'DEV 1 (d1)': {
                ipOrDomain: '172.30.99.19',
                aliases: 'plt-d1-navigator.dev.igt.com plt-d1-bo.dev.igt.com',
            },
            'DEV 5 (d5)': {
                ipOrDomain: '172.30.99.124',
                aliases: 'plt-d5-navigator.dev.igt.com plt-d5-bo.dev.igt.com',
            },
            'I1 integration': {
                ipOrDomain: '172.30.99.44',
                aliases: 'plt-i1-navigator.dev.igt.com plt-i1-bo.dev.igt.com'
            }
        },
        apiProxyTarget: {
            default: {
                'domain': 'https://plt-d1-navigator.dev.igt.com',
                'ip': ''
            },
            other: {
                'D1': {
                    'domain': 'https://plt-d1-navigator.dev.igt.com',
                    'ip': ''
                },
                'D5': {
                    'domain': 'https://plt-d5-navigator.dev.igt.com',
                    'ip': ''
                },
                'I1 integration': {
                    'domain': 'https://plt-i1-navigator.dev.igt.com',
                    'ip': ''
                }
            }
        }
    }),
    //#endregion
};
//#region business modules
exports.NavigatorBusinessModuleObj = {
    'es-common': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightyellow',
        description: 'ui module for all other business modules'
    }),
    'es-global-ui': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightslategray',
        description: 'global helpers for all busines modules and apps'
    }),
    'es-rs-ui': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightcoral',
    }),
    'es-ips-ui': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightblue'
    }),
    'es-olpm-ui': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightcyan'
    }),
    'es-ps-ui': new tools_backend_1.Tools.NavigatorBusinessModule({
        color: 'lightseagreen'
    }),
};
//#endregion
//#region calculated objects
exports.NavigatorBaselineForks = Object.values(NavigatorBaselineForksObject);
exports.NavigatorBaselineForksObj = NavigatorBaselineForksObject;
var colors = {};
Object.keys(exports.NavigatorBusinessModuleObj).forEach(function (moduleName) {
    colors[moduleName] = tools_backend_1.Tools.Colors()[exports.NavigatorBusinessModuleObj[moduleName].color];
});
Object.keys(NavigatorBaselineForksObject).forEach(function (moduleName) {
    colors[moduleName] = tools_backend_1.Tools.Colors()[NavigatorBaselineForksObject[moduleName].color];
});
exports.AllDefaultColors = colors;
//#endregion
