// tnp-ui/recrusive-menu/recrusive-menu.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var RecrusiveMenuComponent = /** @class */ (function () {
    function RecrusiveMenuComponent() {
        this.items = [];
        this.childrenProperty = 'children';
        this.nameProperty = 'name';
        this.filter = {
            props: [],
            values: [],
            levels: []
        };
        this.level = 0;
        this.selected = new core_1.EventEmitter();
    }
    Object.defineProperty(RecrusiveMenuComponent.prototype, "cfilter", {
        get: function () {
            return this.filter.levels.includes(this.level) ? this.filter : {
                props: [],
                values: []
            };
        },
        enumerable: true,
        configurable: true
    });
    RecrusiveMenuComponent.prototype.onSelected = function (item) {
        this.selectedItem = item;
        this.selected.next(item);
    };
    RecrusiveMenuComponent.prototype.ngOnInit = function () {
        // console.log(this.filter)
        // debugger
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], RecrusiveMenuComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "selectedItem", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "childrenProperty", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "nameProperty", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "filter", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "level", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", Object)
    ], RecrusiveMenuComponent.prototype, "selected", void 0);
    RecrusiveMenuComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-recrusive-menu',
            templateUrl: './recrusive-menu.component.html',
            styleUrls: ['./recrusive-menu.component.scss']
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], RecrusiveMenuComponent);
    return RecrusiveMenuComponent;
}());
exports.RecrusiveMenuComponent = RecrusiveMenuComponent;
